<!DOCTYPE html><html lang="en" class="no-js"><head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="MPI, Genomics Aotearoa &amp; NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/hts_workshop_mpi/level2/14_nesi_environment/">
      
      
        <link rel="prev" href="../13_shell_redirection/">
      
      
        <link rel="next" href="../21_assembly_de_novo/">
      
      
      <link rel="icon" href="../../theme_images/icon_fig.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.21">
    
    
      
        <title>1.4 - The NeSI computing environment - High Throughput Sequencing</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2a3383ac.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&amp;display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"><script src="../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#14-the-nesi-computing-environment" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="High Throughput Sequencing" class="md-header__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            High Throughput Sequencing
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              1.4 - The NeSI computing environment
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to light mode" type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to dark mode" type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to system preference" type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../level1/11_nesi_connection/" class="md-tabs__link">
          
  
  
  Level1

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../11_shell_manipulation/" class="md-tabs__link">
          
  
  
  Level2

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../proficiency/2023_proficiency/" class="md-tabs__link">
          
  
  
  Proficiency

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../supplementary/common_terms/" class="md-tabs__link">
          
  
  
  Supplementary

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="High Throughput Sequencing" class="md-nav__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    High Throughput Sequencing
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2">
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Level1
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Level1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/11_nesi_connection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Connecting to NeSI
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/12_shell_navigation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Introducing the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/13_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/21_illumina_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Assesing Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/22_illumina_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Filtering Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/31_nanopore_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Assesing Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/32_nanopore_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/41_blastn_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Annotating with BLAST
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/42_slurm_introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Introduction to slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/43_blast_interpretation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Level2
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Level2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../11_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../12_shell_variables/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Working with variables and loops
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../13_shell_redirection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Working with redirection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#persistant-and-temporary-storage-locations" class="md-nav__link">
    <span class="md-ellipsis">
      Persistant and temporary storage locations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#search-and-load-software-with-the-module-system" class="md-nav__link">
    <span class="md-ellipsis">
      Search and load software with the module system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#considerations-when-working-with-modules" class="md-nav__link">
    <span class="md-ellipsis">
      Considerations when working with modules
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conflicting-dependencies-between-software" class="md-nav__link">
    <span class="md-ellipsis">
      Conflicting dependencies between software
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../21_assembly_de_novo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Genome Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../22_assembly_spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Assembling short-read Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../23_assembly_flye/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Assembling Oxford Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../24_assembly_evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Evaluating the results of an assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../25_assembly_polishing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.5 - Polishing of Oxford Nanopore assemblies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../31_coverage_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.1 - Overview of mapping sequences to a reference genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../32_illumina_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Mapping Illumina sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../33_nanopore_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.3 - Mapping Oxford Nanopore sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../34_mapping_filters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.4 - Filtering and compressing sam files
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../35_mapping_statistics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../41_annotation_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Prediction of protein coding sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../42_annotation_prodigal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Prediction with prodigal
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../43_annotation_augustus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Prediction with AUGUSTUS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../44_annotation_protein/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../45_annotation_kraken2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4">
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proficiency
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Proficiency
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 1)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency_alternate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 2)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5">
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Supplementary
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/common_terms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Common terms
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/fastq_format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The FASTQ file format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/slurm_module_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm module guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#persistant-and-temporary-storage-locations" class="md-nav__link">
    <span class="md-ellipsis">
      Persistant and temporary storage locations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#search-and-load-software-with-the-module-system" class="md-nav__link">
    <span class="md-ellipsis">
      Search and load software with the module system
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#considerations-when-working-with-modules" class="md-nav__link">
    <span class="md-ellipsis">
      Considerations when working with modules
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#conflicting-dependencies-between-software" class="md-nav__link">
    <span class="md-ellipsis">
      Conflicting dependencies between software
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="14-the-nesi-computing-environment">1.4 - The NeSI computing environment<a class="headerlink" href="#14-the-nesi-computing-environment" title="Permanent link">¶</a></h1>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">¶</a></h2>
<div class="admonition clock">
<p class="admonition-title">time</p>
<ul>
<li>Teaching: 30 minutes</li>
</ul>
</div>
<div class="admonition circle-info">
<p class="admonition-title">Objectives and Key points</p>
<h4 id="objectives">Objectives<a class="headerlink" href="#objectives" title="Permanent link">¶</a></h4>
<ul>
<li>Understand the differences between the <code>project/</code> and <code>nobackup/</code> directories.</li>
<li>Understand the key commands when working with the module system.</li>
<li>Understand the key commands when working with <code>slurm</code>.</li>
</ul>
<h4 id="keypoints">Keypoints<a class="headerlink" href="#keypoints" title="Permanent link">¶</a></h4>
<ul>
<li>NeSI provide two data directories for performing bioinformatic work. Data should be kept in the location most appropriate for the task.</li>
<li>The module system is used to load and unload software for specific computing tasks.</li>
<li>Software versioning can be controlled via the module system.</li>
<li>Jobs are submitted to the cluster via the <code>slurm</code>, where they run asynchronously from the users session.</li>
</ul>
</div>
<hr>
<h2 id="persistant-and-temporary-storage-locations">Persistant and temporary storage locations<a class="headerlink" href="#persistant-and-temporary-storage-locations" title="Permanent link">¶</a></h2>
<p>When working on NeSI there are two locations we can use to store our data; the <code>project/</code> directory (where we work for these training workshops) and the <code>nobackup/</code> location. These file locations are similar in terms of how they are accessed:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="nb">cd</span><span class="w"> </span>/nesi/project/nesi03181/
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="nb">cd</span><span class="w"> </span>/nesi/nobackup/nesi03181/
</code></pre></div>
</div>
<p>There are important differences between them. Data on the <code>project/</code> directory is <em>persistent</em>, backed up, and a billed resource. Anything you place in the <code>project/</code> directory will stay there until it is deleted.</p>
<p>In contrast, files written to the <code>nobackup/</code> directory do not last forever. Files that have not been modified (a proxy for 'used') in the last three months are automatically removed from the system. The benefit of this directory though is that we are not charged for our usage of the <code>nobackup/</code> directory.</p>
<p>There are many tools we work with which create a lot of intermediate files as they run, and realistically we do not need to keep everything that produce. In particular, assembly tools such as <code>SPAdes</code> and <code>Canu</code> create a lot of temporary files as they move through rounds of data cleaning, assembly, and refinement but in the end only a handful of the outputs are actually useful for the end user.</p>
<p>When working with large projects, it is best practice to keep your raw data in the <code>project/</code> directory but perform your day to day work in the <code>nobackup/</code> side. Only copy the critical outputs back to the <code>project/</code> side.</p>
<hr>
<h2 id="search-and-load-software-with-the-module-system">Search and load software with the <code>module</code> system<a class="headerlink" href="#search-and-load-software-with-the-module-system" title="Permanent link">¶</a></h2>
<p>Up until this point we have worked exclusively with software that is available by default in <code>bash</code> environments. For biology-related tasks though, these tools alone are not sufficient and we must use specialised software for performing analyses.</p>
<p>While we are free to install new software into NeSI there are many common bioinformatics tools already available on the platform, we just need to know how to access them. The most user-friendly option for finding pre-installed software is the NeSI <a href="https://support.nesi.org.nz/hc/en-gb/sections/360000040076-Supported-Applications">Supported Applications</a> web page. This provides an up to date list of everything available on NeSI. Each piece of software lists the versions installed and links to the software documentation, and every entry is tagged with some handy keywords to enable quick searching.</p>
<p>Alternatively, if we are already logged into NeSI then we can search from the command line to find software relevant to us. To view our currently loaded software modules, we can use the <code>module list</code> command.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>module<span class="w"> </span>list
</code></pre></div>
</div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>Currently Loaded Modules:
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>1) XALT/minimal                      15) imkl-FFTW/2022.0.2-gimpi-2022a      29) libxml2/2.9.10-GCCcore-11.3.0
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>2) NeSI                         (S)  16) gimkl/2022a                         30) libxslt/1.1.34-GCCcore-11.3.0
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>3) slurm                             17) nodejs/16.15.1-GCCcore-11.3.0       31) cURL/7.83.1-GCCcore-11.3.0
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>4) GCCcore/11.3.0                    18) git/2.23.3                          32) netCDF/4.8.1-gimpi-2022a
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>5) zlib/1.2.11-GCCcore-11.3.0        19) ZeroMQ/4.3.4-GCCcore-11.3.0         33) SQLite/3.36.0-GCCcore-11.3.0
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>6) binutils/2.38-GCCcore-11.3.0      20) bzip2/1.0.8-GCCcore-11.3.0          34) Tcl/8.6.10-GCCcore-11.3.0
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>7) GCC/11.3.0                        21) XZ/5.2.5-GCCcore-11.3.0             35) Tk/8.6.10-GCCcore-11.3.0
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>8) libpmi/2-slurm                    22) libpng/1.6.37-GCCcore-11.3.0        36) OpenSSL/1.1.1k-GCCcore-11.3.0
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>9) numactl/2.0.14-GCC-11.3.0         23) freetype/2.11.1-GCCcore-11.3.0      37) Python/3.10.5-gimkl-2022a
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>10) UCX/1.12.1-GCC-11.3.0             24) Szip/2.1.1-GCCcore-11.3.0           38) JupyterLab/.2023.1.0-gimkl-2022a-3.5.3 (H)
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>11) impi/2021.5.1-GCC-11.3.0          25) HDF5/1.12.2-gimpi-2022a             39) craype-broadwell
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a>12) AlwaysIntelMKL/1.0                26) libjpeg-turbo/2.1.3-GCCcore-11.3.0  40) craype-network-infiniband
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>13) imkl/2022.0.2                     27) ncurses/6.2-GCCcore-11.3.0
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a>14) gimpi/2022a                       28) libreadline/8.1-GCCcore-11.3.0
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a>Where:
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a>H:  Hidden Module
</code></pre></div>
</details>
<p>If we want to see what additional software is available to load, there are two options. The first is to simply report a list of every software module available on NeSI:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="c1"># See all tools...</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>module<span class="w"> </span>avail
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="c1"># See all tools that match a given keyword...</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>module<span class="w"> </span>avail<span class="w"> </span>blast
</code></pre></div>
</div>
<details class="success">
<summary>Output (blast version)</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>----------------------------------------------- /opt/nesi/CS400_centos7_bdw/modules/all -----------------------------------------------
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>BLAST/2.3.0                BLAST/2.12.0-GCC-9.2.0         BLASTDB/2023-04                  samblaster/0.1.24-gimkl-2017a
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>BLAST/2.6.0-gimkl-2017a    BLAST/2.13.0-GCC-11.3.0 (D)    BLASTDB/2023-07           (D)    samblaster/0.1.26-GCC-9.2.0   (D)
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>BLAST/2.6.0-gimkl-2018b    BLASTDB/2022-07                RMBlast/2.6.0-gimkl-2017a
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a>BLAST/2.9.0-gimkl-2018b    BLASTDB/2022-10                RMBlast/2.9.0-GCC-7.4.0
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a>BLAST/2.10.0-GCC-9.2.0     BLASTDB/2023-01                RMBlast/2.10.0-GCC-9.2.0  (D)
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a>
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>Where:
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a>D:  Default Module
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>Use "module spider" to find all possible modules.
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a>Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".
</code></pre></div>
</details>
<p>In the first instance, we see everything available on NeSI. In the second, we see everything available on NeSI with the keyword 'blast' in the name. There is also a more thorough search option:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>module<span class="w"> </span>spider<span class="w"> </span>blast
</code></pre></div>
</div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>BLAST:
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>    Description:
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a>    Basic Local Alignment Search Tool, or BLAST, is an algorithm for comparing primary biological sequence information, such as
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a>    the amino-acid sequences of different proteins or the nucleotides of DNA sequences. 
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a>
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a>    Versions:
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>        BLAST/2.3.0
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>        BLAST/2.6.0-gimkl-2017a
<a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>        BLAST/2.6.0-gimkl-2018b
<a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>        BLAST/2.9.0-gimkl-2018b
<a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a>        BLAST/2.10.0-GCC-9.2.0
<a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>        BLAST/2.12.0-GCC-9.2.0
<a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a>        BLAST/2.13.0-GCC-11.3.0
<a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a>
<a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a>For detailed information about a specific "BLAST" module (including how to load the modules) use the module's full name.
<a id="__codelineno-6-19" name="__codelineno-6-19" href="#__codelineno-6-19"></a>For example:
<a id="__codelineno-6-20" name="__codelineno-6-20" href="#__codelineno-6-20"></a>
<a id="__codelineno-6-21" name="__codelineno-6-21" href="#__codelineno-6-21"></a>    $ module spider BLAST/2.9.0-gimkl-2018b
<a id="__codelineno-6-22" name="__codelineno-6-22" href="#__codelineno-6-22"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-23" name="__codelineno-6-23" href="#__codelineno-6-23"></a>
<a id="__codelineno-6-24" name="__codelineno-6-24" href="#__codelineno-6-24"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-25" name="__codelineno-6-25" href="#__codelineno-6-25"></a>BLASTDB:
<a id="__codelineno-6-26" name="__codelineno-6-26" href="#__codelineno-6-26"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-27" name="__codelineno-6-27" href="#__codelineno-6-27"></a>    Description:
<a id="__codelineno-6-28" name="__codelineno-6-28" href="#__codelineno-6-28"></a>    BLAST databases downloaded from NCBI.
<a id="__codelineno-6-29" name="__codelineno-6-29" href="#__codelineno-6-29"></a>
<a id="__codelineno-6-30" name="__codelineno-6-30" href="#__codelineno-6-30"></a>    Versions:
<a id="__codelineno-6-31" name="__codelineno-6-31" href="#__codelineno-6-31"></a>        BLASTDB/2022-07
<a id="__codelineno-6-32" name="__codelineno-6-32" href="#__codelineno-6-32"></a>        BLASTDB/2022-10
<a id="__codelineno-6-33" name="__codelineno-6-33" href="#__codelineno-6-33"></a>        BLASTDB/2023-01
<a id="__codelineno-6-34" name="__codelineno-6-34" href="#__codelineno-6-34"></a>        BLASTDB/2023-04
<a id="__codelineno-6-35" name="__codelineno-6-35" href="#__codelineno-6-35"></a>        BLASTDB/2023-07
<a id="__codelineno-6-36" name="__codelineno-6-36" href="#__codelineno-6-36"></a>
<a id="__codelineno-6-37" name="__codelineno-6-37" href="#__codelineno-6-37"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-38" name="__codelineno-6-38" href="#__codelineno-6-38"></a>For detailed information about a specific "BLASTDB" module (including how to load the modules) use the module's full name.
<a id="__codelineno-6-39" name="__codelineno-6-39" href="#__codelineno-6-39"></a>For example:
<a id="__codelineno-6-40" name="__codelineno-6-40" href="#__codelineno-6-40"></a>
<a id="__codelineno-6-41" name="__codelineno-6-41" href="#__codelineno-6-41"></a>    $ module spider BLASTDB/2023-07
<a id="__codelineno-6-42" name="__codelineno-6-42" href="#__codelineno-6-42"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-43" name="__codelineno-6-43" href="#__codelineno-6-43"></a>
<a id="__codelineno-6-44" name="__codelineno-6-44" href="#__codelineno-6-44"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-45" name="__codelineno-6-45" href="#__codelineno-6-45"></a>RMBlast:
<a id="__codelineno-6-46" name="__codelineno-6-46" href="#__codelineno-6-46"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-47" name="__codelineno-6-47" href="#__codelineno-6-47"></a>    Description:
<a id="__codelineno-6-48" name="__codelineno-6-48" href="#__codelineno-6-48"></a>    RMBlast supports RepeatMasker searches by adding a few necessary features to the stock NCBI blastn program. These include:
<a id="__codelineno-6-49" name="__codelineno-6-49" href="#__codelineno-6-49"></a>    Support for custom matrices ( without KA-Statistics ). Support for cross_match-like complexity adjusted scoring. Cross_match
<a id="__codelineno-6-50" name="__codelineno-6-50" href="#__codelineno-6-50"></a>    is Phil Green's seeded smith-waterman search algorithm. Support for cross_match-like masklevel filtering.. 
<a id="__codelineno-6-51" name="__codelineno-6-51" href="#__codelineno-6-51"></a>
<a id="__codelineno-6-52" name="__codelineno-6-52" href="#__codelineno-6-52"></a>    Versions:
<a id="__codelineno-6-53" name="__codelineno-6-53" href="#__codelineno-6-53"></a>        RMBlast/2.6.0-gimkl-2017a
<a id="__codelineno-6-54" name="__codelineno-6-54" href="#__codelineno-6-54"></a>        RMBlast/2.9.0-GCC-7.4.0
<a id="__codelineno-6-55" name="__codelineno-6-55" href="#__codelineno-6-55"></a>        RMBlast/2.10.0-GCC-9.2.0
<a id="__codelineno-6-56" name="__codelineno-6-56" href="#__codelineno-6-56"></a>
<a id="__codelineno-6-57" name="__codelineno-6-57" href="#__codelineno-6-57"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-58" name="__codelineno-6-58" href="#__codelineno-6-58"></a>For detailed information about a specific "RMBlast" module (including how to load the modules) use the module's full name.
<a id="__codelineno-6-59" name="__codelineno-6-59" href="#__codelineno-6-59"></a>For example:
<a id="__codelineno-6-60" name="__codelineno-6-60" href="#__codelineno-6-60"></a>
<a id="__codelineno-6-61" name="__codelineno-6-61" href="#__codelineno-6-61"></a>    $ module spider RMBlast/2.9.0-GCC-7.4.0
<a id="__codelineno-6-62" name="__codelineno-6-62" href="#__codelineno-6-62"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-63" name="__codelineno-6-63" href="#__codelineno-6-63"></a>
<a id="__codelineno-6-64" name="__codelineno-6-64" href="#__codelineno-6-64"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-65" name="__codelineno-6-65" href="#__codelineno-6-65"></a>samblaster:
<a id="__codelineno-6-66" name="__codelineno-6-66" href="#__codelineno-6-66"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-67" name="__codelineno-6-67" href="#__codelineno-6-67"></a>    Description:
<a id="__codelineno-6-68" name="__codelineno-6-68" href="#__codelineno-6-68"></a>    samblaster is a fast and flexible program for marking duplicates in read-id grouped paired-end SAM files. It can also
<a id="__codelineno-6-69" name="__codelineno-6-69" href="#__codelineno-6-69"></a>    optionally output discordant read pairs and/or split read mappings to separate SAM files, and/or unmapped/clipped reads to a
<a id="__codelineno-6-70" name="__codelineno-6-70" href="#__codelineno-6-70"></a>    separate FASTQ file. When marking duplicates, samblaster will require approximately 20MB of memory per 1M read pairs. 
<a id="__codelineno-6-71" name="__codelineno-6-71" href="#__codelineno-6-71"></a>
<a id="__codelineno-6-72" name="__codelineno-6-72" href="#__codelineno-6-72"></a>    Versions:
<a id="__codelineno-6-73" name="__codelineno-6-73" href="#__codelineno-6-73"></a>        samblaster/0.1.24-gimkl-2017a
<a id="__codelineno-6-74" name="__codelineno-6-74" href="#__codelineno-6-74"></a>        samblaster/0.1.26-GCC-9.2.0
<a id="__codelineno-6-75" name="__codelineno-6-75" href="#__codelineno-6-75"></a>
<a id="__codelineno-6-76" name="__codelineno-6-76" href="#__codelineno-6-76"></a>-----------------------------------------------------------------------------------------------------------------------------------
<a id="__codelineno-6-77" name="__codelineno-6-77" href="#__codelineno-6-77"></a>For detailed information about a specific "samblaster" module (including how to load the modules) use the module's full name.
<a id="__codelineno-6-78" name="__codelineno-6-78" href="#__codelineno-6-78"></a>For example:
<a id="__codelineno-6-79" name="__codelineno-6-79" href="#__codelineno-6-79"></a>
<a id="__codelineno-6-80" name="__codelineno-6-80" href="#__codelineno-6-80"></a>    $ module spider samblaster/0.1.26-GCC-9.2.0
<a id="__codelineno-6-81" name="__codelineno-6-81" href="#__codelineno-6-81"></a>-----------------------------------------------------------------------------------------------------------------------------------
</code></pre></div>
</details>
<p>The difference here is that <code>avail</code> searches the module names, and <code>spider</code> searches their description and other information for the keyword. Both are useful, and in both cases the search is case insensitive. For example, above we used the lowercase spelling of <code>BLAST</code> but in the results we have a mixture of cases (<code>BLAST/2.3.0</code>, <code>RMBlast/2.6.0-gimkl-2017a</code>, <code>samblaster/0.1.24-gimkl-2017a</code>).</p>
<p>When we want to go and load a module, the <code>module load</code> command <strong><em>is case sensitive</em></strong> so we must use the exact result from <code>module avail</code> or <code>module spider</code>.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>module<span class="w"> </span>load<span class="w"> </span>blast/2.3.0
</code></pre></div>
</div>
<details class="failure">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>Lmod has detected the following error:  The following module(s) are unknown: "blast/2.3.0"
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>Please check the spelling or version number. Also try "module spider ..."
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>It is also possible your cache file is out-of-date try:
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>$ module --ignore-cache load "blast/2.3.0"
</code></pre></div>
</details>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>module<span class="w"> </span>load<span class="w"> </span>BLAST/2.3.0
</code></pre></div>
</div>
<p>When we load a module, if there is no feedback from the command prompt then the load was successful. We can now access our tool from the command line like we can for the native tools like <code>grep</code>, <code>ls</code>, and <code>cp</code>.</p>
<hr>
<h2 id="considerations-when-working-with-modules">Considerations when working with modules<a class="headerlink" href="#considerations-when-working-with-modules" title="Permanent link">¶</a></h2>
<p>Software modules are used for a number of reasons. The first is that whenever a session starts (i.e. you log into NeSI) all required tools must be found and loaded by the operating system. With the number of tools available on NeSI this would be prohibitive.</p>
<p>The second reason is that software versions change through time, as new features are added or bugs are fixed. For the <code>samtools</code> software, which we will use later in this training program, there are currently 10 versions of the software installed into NeSI (0.1.18, 0.1.19, 1.3.1, 1.8, 1.9, 1.10, 1.12, 1.13, 1.15, 1.16). All of these are executed through the <code>samtools</code> command, so if all were simultaneously available NeSI would not know which one to use.</p>
<p>In most situations we would want to be working with the most recent version of the software, and if we use the <code>module load</code> command without providing a version number then this is what will load, but sometimes there are reasons to use an older version.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>module<span class="w"> </span>avail<span class="w"> </span>samtools
</code></pre></div>
</div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>----------------------------------------------- /opt/nesi/CS400_centos7_bdw/modules/all -----------------------------------------------
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>SAMtools/0.1.18-gimkl-2017a      SAMtools/0.1.19-GCCcore-11.3.0    SAMtools/1.8-gimkl-2018b    SAMtools/1.12-GCC-9.2.0
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a>SAMtools/0.1.18-gimkl-2018b      SAMtools/0.1.19-gimkl-2017a       SAMtools/1.9-GCC-7.4.0      SAMtools/1.13-GCC-9.2.0
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a>SAMtools/0.1.19-GCCcore-7.4.0    SAMtools/1.3.1-gimkl-2017a        SAMtools/1.10-GCC-9.2.0     SAMtools/1.15.1-GCC-11.3.0
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a>SAMtools/0.1.19-GCCcore-9.2.0    SAMtools/1.8-gimkl-2017a          SAMtools/1.11-GCC-7.4.0     SAMtools/1.16.1-GCC-11.3.0 (D)
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a>
<a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a>Where:
<a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a>D:  Default Module
<a id="__codelineno-11-9" name="__codelineno-11-9" href="#__codelineno-11-9"></a>
<a id="__codelineno-11-10" name="__codelineno-11-10" href="#__codelineno-11-10"></a>Use "module spider" to find all possible modules.
<a id="__codelineno-11-11" name="__codelineno-11-11" href="#__codelineno-11-11"></a>Use "module keyword key1 key2 ..." to search for all possible modules matching any of the "keys".
</code></pre></div>
</details>
<div class="admonition note">
<p class="admonition-title">Be specific in your versioning</p>
<p>It is always better to load a specific version as you then have a record of which version was used should you ever need to revisit your work. This helps to avoid dependency clashes, which is the final consideration when loading modules.</p>
</div>
<hr>
<h2 id="conflicting-dependencies-between-software">Conflicting dependencies between software<a class="headerlink" href="#conflicting-dependencies-between-software" title="Permanent link">¶</a></h2>
<p>One major benefit of working with the module system is that it allows us to avoid conflicts between tools which were written in different programming environments. When we write software, it is very rare to write the entire program from scratch. There are a wealth of publicly available resources which can be used when developing a tool which saves the developer from writing every single line of code they need to achieve their intent. This saves time and leads to more stable and robust code but can be a problem if we are trying to use tools with incompatible dependencies.</p>
<p>For a common example of clashing dependencies, we can try to load the <code>BLAST/2.6.0-gimkl-2017a</code> and <code>SAMtools/1.8-gimkl-2018b</code> modules in the same NeSI session.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>module<span class="w"> </span>load<span class="w"> </span>BLAST/2.6.0-gimkl-2017a
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>module<span class="w"> </span>load<span class="w"> </span>
</code></pre></div>
</div>
<details class="failure">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>Lmod has detected the following error:  These module(s) exist but cannot be loaded as requested: "gimkl/2018b"
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>Try: "module spider gimkl/2018b" to see how to load the module(s).
</code></pre></div>
</details>
<p>This cannot be done. The first module can be loaded without issue, but the second one will always produce an error on load. For example;</p>
<p>This issue occurs because these two different tools were created using a different set of development tools, and both sets of code cannot be active in parallel.</p>
<p>This is one of the key considerations we must keep in mind when working with NeSI.</p>
<p>Ideally, our sessions will be minimalistic and only load a single module for each job, as in order to keep our resource usage minimal and efficient we will use resource requests tailored for the specific job. When we <strong><em>need</em></strong> to perform multiple commands in a single script we must make sure that all modules can be loaded together, or make use of the <code>module purge</code> command to isolate the software at each step of the script.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>module<span class="w"> </span>load<span class="w"> </span>BLAST/2.6.0-gimkl-2017a
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="c1"># Do some BLAST work...</span>
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>
<a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a>module<span class="w"> </span>purge
<a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a>module<span class="w"> </span>load<span class="w"> </span>SAMtools/1.8-gimkl-2018b
<a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a><span class="c1"># Do some samtools work...</span>
</code></pre></div>
</div>
<p>The NeSI module nomenclature works in the manner of <code>SOFTWARE/VERSION-BUILD</code>, for the <code>BLASTn</code> example above we can see this as:</p>
<div class="admonition info">
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>BLAST/2.6.0-gimkl-2017a
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>|-----|-----|----------
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>Tool  Ver.  BUILD
</code></pre></div>
</div>
<p>If tools have the same build description then they should be able to load together, but <strong><em>always test that your modules are compatible by loading them from your JupyterHub terminal before loading them in a script</em>.</strong></p>
<hr>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../13_shell_redirection/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 1.3 - Working with redirection">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                1.3 - Working with redirection
              </div>
            </div>
          </a>
        
        
          
          <a href="../21_assembly_de_novo/" class="md-footer__link md-footer__link--next" aria-label="Next: 2.1 - Genome Assembly">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                2.1 - Genome Assembly
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Intermediate Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate", "navigation.tabs", "navigation.tabs.sticky"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>