<!DOCTYPE html><html lang="en" class="no-js"><head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="MPI, Genomics Aotearoa &amp; NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/hts_workshop_mpi/level2/44_annotation_protein/">
      
      
        <link rel="prev" href="../43_annotation_augustus/">
      
      
        <link rel="next" href="../45_annotation_kraken2/">
      
      
      <link rel="icon" href="../../theme_images/icon_fig.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.21">
    
    
      
        <title>4.4 - Protein annotation with BLAST-like methods - High Throughput Sequencing</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2a3383ac.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&amp;display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"><script src="../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#44-protein-annotation-with-blast-like-methods" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="High Throughput Sequencing" class="md-header__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            High Throughput Sequencing
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              4.4 - Protein annotation with BLAST-like methods
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to light mode" type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to dark mode" type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to system preference" type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../level1/11_nesi_connection/" class="md-tabs__link">
          
  
  
  Level1

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../11_shell_manipulation/" class="md-tabs__link">
          
  
  
  Level2

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../proficiency/2023_proficiency/" class="md-tabs__link">
          
  
  
  Proficiency

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../supplementary/common_terms/" class="md-tabs__link">
          
  
  
  Supplementary

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="High Throughput Sequencing" class="md-nav__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    High Throughput Sequencing
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2">
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Level1
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Level1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/11_nesi_connection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Connecting to NeSI
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/12_shell_navigation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Introducing the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/13_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/21_illumina_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Assesing Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/22_illumina_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Filtering Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/31_nanopore_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Assesing Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/32_nanopore_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/41_blastn_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Annotating with BLAST
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/42_slurm_introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Introduction to slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/43_blast_interpretation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Level2
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Level2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../11_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../12_shell_variables/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Working with variables and loops
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../13_shell_redirection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Working with redirection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../14_nesi_environment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../21_assembly_de_novo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Genome Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../22_assembly_spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Assembling short-read Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../23_assembly_flye/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Assembling Oxford Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../24_assembly_evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Evaluating the results of an assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../25_assembly_polishing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.5 - Polishing of Oxford Nanopore assemblies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../31_coverage_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.1 - Overview of mapping sequences to a reference genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../32_illumina_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Mapping Illumina sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../33_nanopore_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.3 - Mapping Oxford Nanopore sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../34_mapping_filters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.4 - Filtering and compressing sam files
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../35_mapping_statistics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../41_annotation_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Prediction of protein coding sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../42_annotation_prodigal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Prediction with prodigal
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../43_annotation_augustus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Prediction with AUGUSTUS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#refresher-on-the-blast-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      Refresher on the BLAST algorithm
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotating-proteins-using-blastp" class="md-nav__link">
    <span class="md-ellipsis">
      Annotating proteins using BLASTp
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#accelerating-protein-searches-using-diamond" class="md-nav__link">
    <span class="md-ellipsis">
      Accelerating protein searches using diamond
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#comparing-the-outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Comparing the outputs
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../45_annotation_kraken2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4">
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proficiency
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Proficiency
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 1)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency_alternate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 2)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5">
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Supplementary
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/common_terms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Common terms
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/fastq_format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The FASTQ file format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/slurm_module_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm module guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#refresher-on-the-blast-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      Refresher on the BLAST algorithm
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotating-proteins-using-blastp" class="md-nav__link">
    <span class="md-ellipsis">
      Annotating proteins using BLASTp
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#accelerating-protein-searches-using-diamond" class="md-nav__link">
    <span class="md-ellipsis">
      Accelerating protein searches using diamond
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#comparing-the-outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Comparing the outputs
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="44-protein-annotation-with-blast-like-methods">4.4 - Protein annotation with <code>BLAST</code>-like methods<a class="headerlink" href="#44-protein-annotation-with-blast-like-methods" title="Permanent link">¶</a></h1>
<div class="admonition clock">
<p class="admonition-title">time</p>
<ul>
<li>Teaching: 20 minutes</li>
<li>Teaching: 30 minutes</li>
</ul>
</div>
<div class="admonition circle-info">
<p class="admonition-title">Objectives and Key points</p>
<h4 id="objectives">Objectives<a class="headerlink" href="#objectives" title="Permanent link">¶</a></h4>
<ul>
<li>Know how to run a protein sequence classification using <code>BLASTp</code></li>
<li>Know how to run a protein sequence classification using <code>diamond</code></li>
<li>Understand the differences between these tools and when one might be more appropriate than the other.</li>
</ul>
</div>
<hr>
<h2 id="refresher-on-the-blast-algorithm">Refresher on the <code>BLAST</code> algorithm<a class="headerlink" href="#refresher-on-the-blast-algorithm" title="Permanent link">¶</a></h2>
<p>We have covered the major concepts behind the <code>BLAST</code> approach to sequence classification in level 1 training (<a href="../../level1/41_blastn_annotation/">materials here</a>) so will not be covering this in detail today.</p>
<p>However, as a general refresher, remember that the <code>BLAST</code> algorithm approaches sequence classification by comparing small sections of a query sequence to a reference database, and extending the matches where they are found.</p>
<details class="note">
<summary>Review of <code>BLAST</code> steps</summary>
<p><a class="glightbox" data-type="image" data-width="auto" data-height="auto" href="../../img/41_blast_step1.png" data-desc-position="bottom"><img alt="image" src="../../img/41_blast_step1.png" width="600"></a></p>
<p>Where matches are found, BLAST then extends the ends of the seed one position at a time and assesses how well the seed continues to match the targets.  </p>
<p><a class="glightbox" data-type="image" data-width="auto" data-height="auto" href="../../img/41_blast_step2.png" data-desc-position="bottom"><img alt="image" src="../../img/41_blast_step2.png" width="600"></a></p>
<p>Matches and mismatches are recorded and the seed extenstion continues.    </p>
<p><a class="glightbox" data-type="image" data-width="auto" data-height="auto" href="../../img/41_blast_step3.png" data-desc-position="bottom"><img alt="image" src="../../img/41_blast_step3.png" width="600"></a> </p>
<p>BLAST is also able to introduce insertions to preserve a match between query and target sequences.  </p>
<p><a class="glightbox" data-type="image" data-width="auto" data-height="auto" href="../../img/41_blast_step4.png" data-desc-position="bottom"><img alt="image" src="../../img/41_blast_step4.png" width="600"></a> </p>
</details>
<p>In the level 1 training we applied this to nucleotide sequence data, and in practice there is very little difference <code>BLAST</code>-ing against nucleotide or protein sequences.</p>
<details class="question">
<summary>What's the difference?</summary>
<p>Behind the scenes, <code>BLAST</code> uses different seed sizes for its initial matching, as protein sequences are typically shorter than nucleotide sequences and are more information-rich.</p>
<p>There is also a different weighting system applied when evaluating the match between query and target positions to account for whether or not the amino acid residues being compared are identical, or if their chemistry is similar. Similar to the case with <code>BLASTn</code> the scoring system is based on real-world observations of amino acid substitution rates, but is more complex to allow for the differing charge, size, and polarity of the amino acids.</p>
</details>
<hr>
<h2 id="annotating-proteins-using-blastp">Annotating proteins using <code>BLASTp</code><a class="headerlink" href="#annotating-proteins-using-blastp" title="Permanent link">¶</a></h2>
<p>As already mentioned, performing <code>BLAST</code> for protein sequences is very similar to the case for nucleotide sequences. The two differences are that we need to substitute the <code>blastn</code> exectuable with the <code>blastp</code> version, and point out command towards a reference database of protein sequences. We still load the same modules are for a nucleotide <code>BLAST</code>, there are just a few parts of the <code>slurm</code> script which need to change from the last time we ran this job.</p>
<div class="admonition info">
<p class="admonition-title">Using BLAST databases</p>
<p>If you recall the level 1 traning, you will remember that NeSI periodically download and store NCBI <code>BLAST</code> databases in a common location, accessed through the <code>BLASTDB/YYYY-MM</code> module series.</p>
<p>Protein <code>BLAST</code> searches are significantly slower than nucleotide searches, so we will not be using this resource today as the queue and run times for the jobs we would need to run are not practical for the workshop today. We will instead use a local version of the <a href="https://www.uniprot.org/">UniProt</a> <code>SwissProt</code> database which is much smaller.</p>
</div>
<p>Navigate to the <code>/nesi/project/nesi03181/phel/&lt;username&gt;/level2/annotation_protein/</code> directory and prepare the following <code>slurm</code> script:</p>
<div class="admonition file-code">
<p class="admonition-title">level2_blast.sl</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="c1">#SBATCH --account       nesi03181</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="c1">#SBATCH --job-name      level2_blast</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="c1">#SBATCH --time          00:20:00</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="c1">#SBATCH --cpus-per-task 10</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="c1">#SBATCH --mem           20G</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="c1">#SBATCH --error         level2_blast.%j.err</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="c1">#SBATCH --output        level2_blast.%j.out</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="c1">#SBATCH --mail-type     END</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="c1">#SBATCH --mail-user     YOUR_EMAIL</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a>module<span class="w"> </span>purge
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a>module<span class="w"> </span>load<span class="w"> </span>BLAST/2.13.0-GCC-11.3.0
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="nb">cd</span><span class="w"> </span>/nesi/project/nesi03181/phel/&lt;username&gt;/level2/annotation_protein/
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a>blastp<span class="w"> </span>-num_threads<span class="w"> </span><span class="si">${</span><span class="nv">SLURM_CPUS_PER_TASK</span><span class="si">}</span><span class="w"> </span>-max_target_seqs<span class="w"> </span><span class="m">10</span><span class="w"> </span>-evalue<span class="w"> </span>1e-3<span class="w"> </span>-outfmt<span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="w">    </span>-db<span class="w"> </span>/nesi/project/nesi03181/phel/databases/swissprot_blastp/uniprot_sprot<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="w">    </span>-query<span class="w"> </span>input/input_seqs.faa<span class="w"> </span>-out<span class="w"> </span>outputs/blastp.txt
</code></pre></div>
</div>
<p>Submit this job to <code>slurm</code>:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>sbatch<span class="w"> </span>level2_blast.sl
</code></pre></div>
</div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>Submitted batch job XXXXXXXX
</code></pre></div>
</details>
<hr>
<h2 id="accelerating-protein-searches-using-diamond">Accelerating protein searches using <code>diamond</code><a class="headerlink" href="#accelerating-protein-searches-using-diamond" title="Permanent link">¶</a></h2>
<p>Given the time required to produce results with the <code>BLASTp</code> approach, a lot of work has been conducted to speed up the results of protein alignment searches. A very useful alternative to <code>BLAST</code> when working with protein data, or when trying to perform translated (nucleotide queries, protein targets) searches is <code>diamond</code> (<a href="https://doi.org/10.1038/s41592-021-01101-x">Buchfink <em>et al.</em>, 2021</a>)</p>
<div class="admonition info">
<p class="admonition-title">Differences in <code>diamond</code> versions</p>
<p>In previous releases, <code>diamond</code> gave faster results at the cost of sensitivity but the more recent release of the tool purports comparable sensitivity with <code>BLASTp</code> at a greatly reduced run time. See Figure 1 in the manuscript above for comparisons of sensitivity and speed between more recent releases of <code>diamond</code> and <code>BLASTp</code>.</p>
</div>
<details class="warning">
<summary>Producing a <code>diamond</code> database</summary>
<p>Unfortunately, <code>diamond</code> requires us to create our own classification database - there is no NCBI-supported release. However, <code>diamond</code> has a built-in method to modify an existing <code>BLAST</code> database to be compatible with <code>diamond</code>.</p>
<p>You will probably not ever need to perform this operation yourself, but the database we are using today was produced using the following commands:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>module<span class="w"> </span>purge
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>module<span class="w"> </span>load<span class="w"> </span>BLASTDB/2023-10<span class="w"> </span>DIAMOND/2.1.6-GCC-11.3.0
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>mkdir<span class="w"> </span>/nesi/project/nesi03181/phel/databases/diamond_nr/
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="nb">cd</span><span class="w"> </span>/nesi/project/nesi03181/phel/databases/diamond_nr/
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>cp<span class="w"> </span><span class="si">${</span><span class="nv">BLASTDB</span><span class="si">}</span>/nr*<span class="w"> </span>diamond_nr/
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>diamond<span class="w"> </span>prepdb<span class="w"> </span>-d<span class="w"> </span>diamond_nr/nr
</code></pre></div>
</div>
</details>
<p>With our database in place, the command for running <code>diamond</code> is very similar (by design) to <code>BLASTp</code>:</p>
<div class="admonition file-code">
<p class="admonition-title">level2_blast.sl</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="c1">#SBATCH --account       nesi03181</span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="c1">#SBATCH --job-name      level2_diamond</span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a><span class="c1">#SBATCH --time          00:10:00</span>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="c1">#SBATCH --cpus-per-task 16</span>
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a><span class="c1">#SBATCH --mem           20G</span>
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a><span class="c1">#SBATCH --error         level2_diamond.%j.err</span>
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a><span class="c1">#SBATCH --output        level2_diamond.%j.out</span>
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a><span class="c1">#SBATCH --mail-type     END</span>
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a><span class="c1">#SBATCH --mail-user     YOUR_EMAIL</span>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a>
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a>module<span class="w"> </span>purge
<a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a>module<span class="w"> </span>load<span class="w"> </span>DIAMOND/2.1.6-GCC-11.3.0
<a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a>
<a id="__codelineno-4-15" name="__codelineno-4-15" href="#__codelineno-4-15"></a><span class="nb">cd</span><span class="w"> </span>/nesi/project/nesi03181/phel/&lt;username&gt;/level2/annotation_protein/
<a id="__codelineno-4-16" name="__codelineno-4-16" href="#__codelineno-4-16"></a>
<a id="__codelineno-4-17" name="__codelineno-4-17" href="#__codelineno-4-17"></a>diamond<span class="w"> </span>blastp<span class="w"> </span>--threads<span class="w"> </span><span class="si">${</span><span class="nv">SLURM_CPUS_PER_TASK</span><span class="si">}</span><span class="w"> </span>--ultra-sensitive<span class="w"> </span>--max-target-seqs<span class="w"> </span><span class="m">10</span><span class="w"> </span>--evalue<span class="w"> </span>1e-3<span class="w"> </span>--outfmt<span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-4-18" name="__codelineno-4-18" href="#__codelineno-4-18"></a><span class="w">    </span>--db<span class="w"> </span>/nesi/project/nesi03181/phel/databases/swissprot_dmnd/uniprot_sprot.dmnd<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-4-19" name="__codelineno-4-19" href="#__codelineno-4-19"></a><span class="w">    </span>--query<span class="w"> </span>input/input_seqs.faa<span class="w"> </span>--out<span class="w"> </span>outputs/diamond.txt
</code></pre></div>
</div>
<p>Submit this job to <code>slurm</code>:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>sbatch<span class="w"> </span>level2_blast.sl
</code></pre></div>
</div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>Submitted batch job XXXXXXXX
</code></pre></div>
</details>
<hr>
<h2 id="comparing-the-outputs">Comparing the outputs<a class="headerlink" href="#comparing-the-outputs" title="Permanent link">¶</a></h2>
<p>In both of the jbos above, we specified the tab-delimited <code>BLAST6</code> output format, which is a table of the form;</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Column header</th>
<th style="text-align: left;">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">qseqid</td>
<td style="text-align: left;">Sequence ID of the query sequence (input file)</td>
</tr>
<tr>
<td style="text-align: left;">sseqid</td>
<td style="text-align: left;">Sequence ID of the target sequence (reference database)</td>
</tr>
<tr>
<td style="text-align: left;">pident</td>
<td style="text-align: left;">Percentage of identical positions between query and target</td>
</tr>
<tr>
<td style="text-align: left;">length</td>
<td style="text-align: left;">Alignment length (sequence overlap) of the common region between query and target</td>
</tr>
<tr>
<td style="text-align: left;">mismatch</td>
<td style="text-align: left;">Number of mismatches between query and target</td>
</tr>
<tr>
<td style="text-align: left;">gapopen</td>
<td style="text-align: left;">Number of gap openings in the alignment</td>
</tr>
<tr>
<td style="text-align: left;">qstart</td>
<td style="text-align: left;">Position in the query sequence where alignment begins</td>
</tr>
<tr>
<td style="text-align: left;">qend</td>
<td style="text-align: left;">Position in the query sequence where alignment ends</td>
</tr>
<tr>
<td style="text-align: left;">sstart</td>
<td style="text-align: left;">Position in the target sequence where alignment begins</td>
</tr>
<tr>
<td style="text-align: left;">send</td>
<td style="text-align: left;">Position in the target sequence where alignment ends</td>
</tr>
<tr>
<td style="text-align: left;">evalue</td>
<td style="text-align: left;">The E-value for the query/target match, as described above</td>
</tr>
<tr>
<td style="text-align: left;">bitscore</td>
<td style="text-align: left;">The bit score for the query/target match, as described above</td>
</tr>
</tbody>
</table>
<p>As there are around 3,000 proteins in our input file which were classified, we are just going to focus on a single sequence and how it's annotation differs between <code>BLASTp</code> and <code>diamond</code>.</p>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>Use your knowledge the command line to print all results for the query sequence <code>46ee54c6-57d2-2d22-f454-9a5737db31e7_1</code> in both output files. You can either keep the results on the command line, or redirect them into files for analysis.</p>
<details class="circle-check">
<summary>Solution</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>grep<span class="w"> </span><span class="s2">"46ee54c6-57d2-2d22-f454-9a5737db31e7_1"</span><span class="w"> </span>outputs/blastp.txt<span class="w"> </span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>grep<span class="w"> </span><span class="s2">"46ee54c6-57d2-2d22-f454-9a5737db31e7_1"</span><span class="w"> </span>outputs/diamond.txt
</code></pre></div>
<details class="success">
<summary>outputs/blastp.txt</summary>
<table>
<thead>
<tr>
<th style="text-align: left;">qseqid</th>
<th style="text-align: left;">sseqid</th>
<th style="text-align: center;">pident</th>
<th style="text-align: center;">length</th>
<th style="text-align: center;">mismatch</th>
<th style="text-align: center;">gapopen</th>
<th style="text-align: center;">qstart</th>
<th style="text-align: center;">qend</th>
<th style="text-align: center;">sstart</th>
<th style="text-align: center;">send</th>
<th style="text-align: center;">evalue</th>
<th style="text-align: center;">bitscore</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P17779|RDRP_PVXX3</td>
<td style="text-align: center;">100.000</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1499</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P09395|RDRP_PVX</td>
<td style="text-align: center;">97.218</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1462</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q07630|RDRP_PVXHB</td>
<td style="text-align: center;">83.449</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">119</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1256</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P22591|RDRP_PVXCP</td>
<td style="text-align: center;">82.476</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">126</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1237</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P20951|RDRP_PMV</td>
<td style="text-align: center;">48.563</td>
<td style="text-align: center;">348</td>
<td style="text-align: center;">174</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">370</td>
<td style="text-align: center;">717</td>
<td style="text-align: center;">699</td>
<td style="text-align: center;">1041</td>
<td style="text-align: center;">2.79e-97</td>
<td style="text-align: center;">332</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P20951|RDRP_PMV</td>
<td style="text-align: center;">41.489</td>
<td style="text-align: center;">188</td>
<td style="text-align: center;">110</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">188</td>
<td style="text-align: center;">230</td>
<td style="text-align: center;">417</td>
<td style="text-align: center;">7.28e-39</td>
<td style="text-align: center;">159</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P09498|RDRP_WCMVM</td>
<td style="text-align: center;">50.140</td>
<td style="text-align: center;">357</td>
<td style="text-align: center;">168</td>
<td style="text-align: center;">6</td>
<td style="text-align: center;">364</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">446</td>
<td style="text-align: center;">793</td>
<td style="text-align: center;">3.73e-96</td>
<td style="text-align: center;">327</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P09498|RDRP_WCMVM</td>
<td style="text-align: center;">40.556</td>
<td style="text-align: center;">180</td>
<td style="text-align: center;">104</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;">193</td>
<td style="text-align: center;">235</td>
<td style="text-align: center;">411</td>
<td style="text-align: center;">4.56e-30</td>
<td style="text-align: center;">131</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P15402|RDRP_WCMVO</td>
<td style="text-align: center;">49.860</td>
<td style="text-align: center;">357</td>
<td style="text-align: center;">169</td>
<td style="text-align: center;">6</td>
<td style="text-align: center;">364</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">446</td>
<td style="text-align: center;">793</td>
<td style="text-align: center;">2.67e-94</td>
<td style="text-align: center;">322</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P15402|RDRP_WCMVO</td>
<td style="text-align: center;">38.776</td>
<td style="text-align: center;">196</td>
<td style="text-align: center;">116</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">210</td>
<td style="text-align: center;">236</td>
<td style="text-align: center;">427</td>
<td style="text-align: center;">1.67e-29</td>
<td style="text-align: center;">129</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q07518|RDRP_P1AMV</td>
<td style="text-align: center;">47.238</td>
<td style="text-align: center;">362</td>
<td style="text-align: center;">180</td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">363</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">531</td>
<td style="text-align: center;">886</td>
<td style="text-align: center;">1.71e-88</td>
<td style="text-align: center;">306</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q07518|RDRP_P1AMV</td>
<td style="text-align: center;">41.765</td>
<td style="text-align: center;">170</td>
<td style="text-align: center;">96</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;">182</td>
<td style="text-align: center;">236</td>
<td style="text-align: center;">403</td>
<td style="text-align: center;">7.82e-31</td>
<td style="text-align: center;">134</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q918W3|RDRP_ICRSV</td>
<td style="text-align: center;">48.986</td>
<td style="text-align: center;">345</td>
<td style="text-align: center;">171</td>
<td style="text-align: center;">5</td>
<td style="text-align: center;">375</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">821</td>
<td style="text-align: center;">1160</td>
<td style="text-align: center;">1.13e-87</td>
<td style="text-align: center;">305</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q918W3|RDRP_ICRSV</td>
<td style="text-align: center;">26.070</td>
<td style="text-align: center;">257</td>
<td style="text-align: center;">175</td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;">268</td>
<td style="text-align: center;">242</td>
<td style="text-align: center;">485</td>
<td style="text-align: center;">1.60e-20</td>
<td style="text-align: center;">100</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P28897|RDRP_SMYEA</td>
<td style="text-align: center;">48.387</td>
<td style="text-align: center;">341</td>
<td style="text-align: center;">170</td>
<td style="text-align: center;">6</td>
<td style="text-align: center;">380</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">493</td>
<td style="text-align: center;">828</td>
<td style="text-align: center;">1.08e-85</td>
<td style="text-align: center;">298</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P28897|RDRP_SMYEA</td>
<td style="text-align: center;">32.524</td>
<td style="text-align: center;">206</td>
<td style="text-align: center;">129</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">211</td>
<td style="text-align: center;">254</td>
<td style="text-align: center;">458</td>
<td style="text-align: center;">1.05e-24</td>
<td style="text-align: center;">114</td>
</tr>
</tbody>
</table>
</details>
<details class="success">
<summary>outputs/diamond.txt</summary>
<table>
<thead>
<tr>
<th style="text-align: left;">qseqid</th>
<th style="text-align: left;">sseqid</th>
<th style="text-align: center;">pident</th>
<th style="text-align: center;">length</th>
<th style="text-align: center;">mismatch</th>
<th style="text-align: center;">gapopen</th>
<th style="text-align: center;">qstart</th>
<th style="text-align: center;">qend</th>
<th style="text-align: center;">sstart</th>
<th style="text-align: center;">send</th>
<th style="text-align: center;">evalue</th>
<th style="text-align: center;">bitscore</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P17779|RDRP_PVXX3</td>
<td style="text-align: center;">100</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1425</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P09395|RDRP_PVX</td>
<td style="text-align: center;">97.2</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">20</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1390</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q07630|RDRP_PVXHB</td>
<td style="text-align: center;">83.4</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">119</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1197</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P22591|RDRP_PVXCP</td>
<td style="text-align: center;">82.5</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">126</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">240</td>
<td style="text-align: center;">958</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">1179</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P20951|RDRP_PMV</td>
<td style="text-align: center;">34.8</td>
<td style="text-align: center;">819</td>
<td style="text-align: center;">425</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">717</td>
<td style="text-align: center;">230</td>
<td style="text-align: center;">1041</td>
<td style="text-align: center;">8.77e-125</td>
<td style="text-align: center;">409</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P09498|RDRP_WCMVM</td>
<td style="text-align: center;">37.3</td>
<td style="text-align: center;">707</td>
<td style="text-align: center;">294</td>
<td style="text-align: center;">12</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">235</td>
<td style="text-align: center;">793</td>
<td style="text-align: center;">5.69e-117</td>
<td style="text-align: center;">384</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P15402|RDRP_WCMVO</td>
<td style="text-align: center;">36.7</td>
<td style="text-align: center;">706</td>
<td style="text-align: center;">298</td>
<td style="text-align: center;">12</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">236</td>
<td style="text-align: center;">793</td>
<td style="text-align: center;">7.69e-115</td>
<td style="text-align: center;">378</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|Q07518|RDRP_P1AMV</td>
<td style="text-align: center;">38.5</td>
<td style="text-align: center;">716</td>
<td style="text-align: center;">365</td>
<td style="text-align: center;">23</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">236</td>
<td style="text-align: center;">886</td>
<td style="text-align: center;">4.72e-114</td>
<td style="text-align: center;">377</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P28897|RDRP_SMYEA</td>
<td style="text-align: center;">34.2</td>
<td style="text-align: center;">710</td>
<td style="text-align: center;">323</td>
<td style="text-align: center;">13</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">717</td>
<td style="text-align: center;">254</td>
<td style="text-align: center;">826</td>
<td style="text-align: center;">5.27e-103</td>
<td style="text-align: center;">346</td>
</tr>
<tr>
<td style="text-align: left;">46ee54c6-57d2-2d22-f454-9a5737db31e7_1</td>
<td style="text-align: left;">sp|P15095|RDRP_NMV</td>
<td style="text-align: center;">32.2</td>
<td style="text-align: center;">860</td>
<td style="text-align: center;">419</td>
<td style="text-align: center;">23</td>
<td style="text-align: center;">15</td>
<td style="text-align: center;">719</td>
<td style="text-align: center;">243</td>
<td style="text-align: center;">1093</td>
<td style="text-align: center;">5.58e-100</td>
<td style="text-align: center;">340</td>
</tr>
</tbody>
</table>
</details>
</details>
</div>
<p>Comparing the outputs, you can see that <code>diamond</code> produced slightly fewer hits than <code>BLASTp</code>, but the top four results for both methods are the same:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Target</th>
<th style="text-align: center;"><code>BLASTp</code><br>Identity</th>
<th style="text-align: center;"><br>E-value</th>
<th style="text-align: center;"><br>Bitscore</th>
<th style="text-align: center;"><code>diamond</code><br>Identity</th>
<th style="text-align: center;"><br>E-value</th>
<th style="text-align: center;"><br>Bitscore</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">sp|P17779|RDRP_PVXX3</td>
<td style="text-align: center;">100.000</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,499</td>
<td style="text-align: center;">100.0</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,425</td>
</tr>
<tr>
<td style="text-align: left;">sp|P09395|RDRP_PVX</td>
<td style="text-align: center;">97.218</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,462</td>
<td style="text-align: center;">97.2</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,390</td>
</tr>
<tr>
<td style="text-align: left;">sp|Q07630|RDRP_PVXHB</td>
<td style="text-align: center;">83.449</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,256</td>
<td style="text-align: center;">83.4</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,197</td>
</tr>
<tr>
<td style="text-align: left;">sp|P22591|RDRP_PVXCP</td>
<td style="text-align: center;">82.476</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,237</td>
<td style="text-align: center;">82.5</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1,179</td>
</tr>
</tbody>
</table>
<p>We can check the complete annotation of this protein to determine its function and organism from which the reference sequence originated.</p>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>Search the top result(s) through the <a href="https://www.uniprot.org/">UniProt search tool</a> and report the most likely function of this sequence, and the organism that it was obtained from.</p>
<details class="circle-check">
<summary>Solution</summary>
<table>
<thead>
<tr>
<th style="text-align: left;">Sequence</th>
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Organism</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">sp|P17779|RDRP_PVXX3</td>
<td style="text-align: left;">RNA-directed RNA polymerase</td>
<td style="text-align: left;">Potato virus X (strain X3)</td>
</tr>
<tr>
<td style="text-align: left;">sp|P09395|RDRP_PVX</td>
<td style="text-align: left;">RNA-directed RNA polymerase</td>
<td style="text-align: left;">Potato virus X</td>
</tr>
<tr>
<td style="text-align: left;">sp|Q07630|RDRP_PVXHB</td>
<td style="text-align: left;">RNA-directed RNA polymerase</td>
<td style="text-align: left;">Potato virus X</td>
</tr>
<tr>
<td style="text-align: left;">sp|P22591|RDRP_PVXCP</td>
<td style="text-align: left;">RNA-directed RNA polymerase</td>
<td style="text-align: left;">Potato virus X (strain CP)</td>
</tr>
</tbody>
</table>
</details>
</div>
<hr>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../43_annotation_augustus/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 4.3 - Prediction with AUGUSTUS">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                4.3 - Prediction with AUGUSTUS
              </div>
            </div>
          </a>
        
        
          
          <a href="../45_annotation_kraken2/" class="md-footer__link md-footer__link--next" aria-label="Next: 4.5 - Classification of sequences with k-mer composition profiles">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                4.5 - Classification of sequences with k-mer composition profiles
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Intermediate Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate", "navigation.tabs", "navigation.tabs.sticky"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>