# Working with Oxford Nanopore data

* Teaching: 20 minutes
* Exercises: 10 minutes

#### Objectives

* Know how to assess the quality of Illumina sequence data using visualization tools such as pycoQC
* Be able to perform quality filtering of nanopore data and to remove short reads

#### Keypoints

* Base calling is in most cases done in parallel with sequencing through the `MinKNOW` tool
* `pycoQC` is a great tool for generating interactive reports to explore key aspects of a Nanopore sequencing dataset
* Sequences can be filtered to remove low quality reads and short reads using `nanofilt`

---

## Contents
1. [Notes on base calling](#notes-on-base-calling)
1. [Assessing sequence quality](#assessing-sequence-quality)
1. [Removing short and low quality reads](#Removing-short-and-low-quality-reads)

---

## Notes on base calling

Base calling is the process of translating the raw signal generated by the Nanopore sequencer into nucleotide bases (*i.e.* ATCG). Base calling is in most cases performed 'live' in real time while sequencing is performed so that when the sequencing run is finished, we have our data ready for analysis. There are several base calling tools available, but the official production version developed by Oxford Nanopore Technologies is `guppy`. `guppy` uses neural networks to convert raw signal into sequence data and in addition can  perform identification and removal of Oxford Nanopore sequencing adapters, barcodes, and low quality sequence regions.

When we set the sequencing run parameters using the `MminKNOW` software, we can specify if we want the base calling to be performed automatically by the computer. When the basecalling option is ON, the output of the sequencing run will be saved in fastq files.

![](../img/02_minknow_enable_basecalling.png)

When performing base calling in `guppy` there are several prediction models that can be used, offering increased sequence quality and accuracy of translation at the cost of longer run times and more computational power. `MinKNOW` allows us to select between the fast and high-accuracy methods.

![](../img/02_minknow_select_model.png)

>**Note:** With the release of `guppy` version 5 there are also 'super accurate' models which provide higher quality translations than the high-accuracy model. These models can only be used when running `guppy` throug the command line.

If we choose to not perform the base calling in real time, we will need to perform the manually call the electronic signal files on the raw output files. These are stored in the `fast5` format. In order to do this, we can work with `guppy` in the command line in NeSI or the computers which PHEL use to run the MinION devices. We are not going to do this in the present workshop, as we normally request minKNOW to perform basecalling in real time.

When performing base calling during the sequencing run, we also have the option to quality filter the sequences straight away, saving us the need to perform this task afterwards. When using this tool, the output fastq files are split into folders of sequences that passed and failed to pass the required Q threshold so we do not *lose* the low quality sequences.

![](../img/02_minknow_quality_filter.png)

![](../img/02_minknow_quality_threshold.png)

---

## Assessing sequence quality

The initial step for every sequencing project is quality control to assess the quality of your data. This will give you some statistics of your sequencing data, such as length and quality score distributions, as well as highlight potential problems with your input DNA/RNA, the sequencing run or the output itself.

An increasing number of tools are available for sequence data assessment, with different strengths and intended applications. In this session we will introduce the tool `pycoQC`. This is a data visualisation and quality control tool for Nanopore data. In contrast to `FastQC` (which we used in the previous exercise) `pycoQC` requires a specific output file from the sequencing run which is **_not_** in fastq format. This file, named `sequencing_summary.txt` is generated `guppy` during run time.. This file will be located in the same folder that contains the output from the base caller.

>**Note:** Although `FastQC` *can* be used for assessing Nanopore data, it is designed for short sequences and the figures and outputs when long reads are visualised can be confusing without customisation.

> ### Exercise
>
> Using your knowledge of the shell, navigate to `/nesi/project/nesi03181/phel/module_2/` and copy the directory `nanopore_data/` to your own working directory (/nesi/project/nesi03181/phel/USERNAME/).
> 
> <details>
> <summary>Solution</summary>
>
> ```bash
> $ cd /nesi/project/nesi03181/phel/module_2/
  $ cp -r nanopore_data/ ../USERNAME/fastq_processing/data/
> ```
> </details>

To activate pycoQC on NeSI, we need to first load the `slurm` module.

```bash
$ module purge
$ module load pycoQC/2.5.2-gimkl-2020a-Python-3.8.2
```

To run `pycoQC` we simply provide the input file and an output file destination.

```bash
$ cd /nesi/project/nesi03181/phel/USERNAME/fastq_processing/
$ pycoQC -f data/nanopore_data/sequencing_summary.txt -o results/pycoQC_report.html
```

`pycoQC`  generates output reports as html files, which we can open the same way as we did for `FastQC`.


---

## Removing short and low quality reads

xxx
