<!DOCTYPE html><html lang="en" class="no-js"><head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="MPI, Genomics Aotearoa &amp; NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/hts_workshop_mpi/level2/35_mapping_statistics/">
      
      
        <link rel="prev" href="../34_mapping_filters/">
      
      
        <link rel="next" href="../41_annotation_overview/">
      
      
      <link rel="icon" href="../../theme_images/icon_fig.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.21">
    
    
      
        <title>3.5 - Summarising coverage information from a mapping file - High Throughput Sequencing</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2a3383ac.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&amp;display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"><script src="../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#35-summarising-coverage-information-from-a-mapping-file" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="High Throughput Sequencing" class="md-header__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            High Throughput Sequencing
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              3.5 - Summarising coverage information from a mapping file
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to light mode" type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to dark mode" type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to system preference" type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../level1/11_nesi_connection/" class="md-tabs__link">
          
  
  
  Level1

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../11_shell_manipulation/" class="md-tabs__link">
          
  
  
  Level2

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../proficiency/2023_proficiency/" class="md-tabs__link">
          
  
  
  Proficiency

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../supplementary/common_terms/" class="md-tabs__link">
          
  
  
  Supplementary

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="High Throughput Sequencing" class="md-nav__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    High Throughput Sequencing
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2">
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Level1
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Level1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/11_nesi_connection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Connecting to NeSI
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/12_shell_navigation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Introducing the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/13_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/21_illumina_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Assesing Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/22_illumina_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Filtering Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/31_nanopore_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Assesing Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/32_nanopore_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/41_blastn_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Annotating with BLAST
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/42_slurm_introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Introduction to slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/43_blast_interpretation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Level2
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Level2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../11_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../12_shell_variables/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Working with variables and loops
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../13_shell_redirection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Working with redirection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../14_nesi_environment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../21_assembly_de_novo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Genome Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../22_assembly_spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Assembling short-read Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../23_assembly_flye/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Assembling Oxford Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../24_assembly_evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Evaluating the results of an assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../25_assembly_polishing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.5 - Polishing of Oxford Nanopore assemblies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../31_coverage_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.1 - Overview of mapping sequences to a reference genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../32_illumina_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Mapping Illumina sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../33_nanopore_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.3 - Mapping Oxford Nanopore sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../34_mapping_filters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.4 - Filtering and compressing sam files
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summarising-the-contents-of-a-bam-file" class="md-nav__link">
    <span class="md-ellipsis">
      Summarising the contents of a bam file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#producing-a-coverage-table-from-a-complete-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Producing a coverage table from a complete alignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#producing-a-coverage-table-from-a-specific-region-of-an-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Producing a coverage table from a specific region of an alignment
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../41_annotation_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Prediction of protein coding sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../42_annotation_prodigal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Prediction with prodigal
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../43_annotation_augustus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Prediction with AUGUSTUS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../44_annotation_protein/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../45_annotation_kraken2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4">
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proficiency
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Proficiency
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 1)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency_alternate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 2)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5">
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Supplementary
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/common_terms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Common terms
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/fastq_format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The FASTQ file format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/slurm_module_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm module guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summarising-the-contents-of-a-bam-file" class="md-nav__link">
    <span class="md-ellipsis">
      Summarising the contents of a bam file
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#producing-a-coverage-table-from-a-complete-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Producing a coverage table from a complete alignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#producing-a-coverage-table-from-a-specific-region-of-an-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Producing a coverage table from a specific region of an alignment
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="35-summarising-coverage-information-from-a-mapping-file">3.5 - Summarising coverage information from a mapping file<a class="headerlink" href="#35-summarising-coverage-information-from-a-mapping-file" title="Permanent link">¶</a></h1>
<div class="admonition clock">
<p class="admonition-title">time</p>
<ul>
<li>Teaching: 15 minutes</li>
<li>Exercises: 20 minutes</li>
</ul>
</div>
<div class="admonition circle-info">
<p class="admonition-title">Objectives and Key points</p>
<h4 id="objectives">Objectives<a class="headerlink" href="#objectives" title="Permanent link">¶</a></h4>
<ul>
<li>Use <code>samtools flagstat</code> to get an overview of a <code>bam</code> file content.</li>
<li>Use <code>samtools depth</code> to produce a coverage table from the complete <code>bam</code> record.</li>
<li>Use <code>samtools depth</code> to produce a coverage table from a subsection of a complete <code>bam</code> record.</li>
</ul>
<h4 id="keypoints">Keypoints<a class="headerlink" href="#keypoints" title="Permanent link">¶</a></h4>
<ul>
<li>Understand the reasons for summarising a <code>bam</code> file without digging into the contents in detail.</li>
<li>Understand the situations in which you might require the coverage information, either in its entirity or for a specific genomic region.</li>
</ul>
</div>
<hr>
<h3 id="summarising-the-contents-of-a-bam-file">Summarising the contents of a <code>bam</code> file<a class="headerlink" href="#summarising-the-contents-of-a-bam-file" title="Permanent link">¶</a></h3>
<p>This section will start with a new mapping file, although you could apply these commands to anything produced as part of the <a href="../34_mapping_filters/">filtering exercises</a> instead. Navigate to the <code>/nesi/project/nesi03181/phel/USERNAME/level2/mapping_statistics/</code> directory and view the contents.</p>
<p>We wil take a look at the high-level summary of the contents of these files with the <code>samtools flagstat</code> command:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>module<span class="w"> </span>purge
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>module<span class="w"> </span>load<span class="w"> </span>SAMtools/1.16.1-GCC-11.3.0
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>samtools<span class="w"> </span>flagstat<span class="w"> </span>SRR18260232.raw_reads.bam
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>8630882 + 0 in total (QC-passed reads + QC-failed reads)
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>8629544 + 0 primary
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>1338 + 0 secondary
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>0 + 0 supplementary
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>0 + 0 duplicates
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>0 + 0 primary duplicates
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>3236663 + 0 mapped (37.50% : N/A)
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>3235325 + 0 primary mapped (37.49% : N/A)
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>8629544 + 0 paired in sequencing
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>4314772 + 0 read1
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>4314772 + 0 read2
<a id="__codelineno-1-12" name="__codelineno-1-12" href="#__codelineno-1-12"></a>3040828 + 0 properly paired (35.24% : N/A)
<a id="__codelineno-1-13" name="__codelineno-1-13" href="#__codelineno-1-13"></a>3049552 + 0 with itself and mate mapped
<a id="__codelineno-1-14" name="__codelineno-1-14" href="#__codelineno-1-14"></a>185773 + 0 singletons (2.15% : N/A)
<a id="__codelineno-1-15" name="__codelineno-1-15" href="#__codelineno-1-15"></a>0 + 0 with mate mapped to a different chr
<a id="__codelineno-1-16" name="__codelineno-1-16" href="#__codelineno-1-16"></a>0 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</code></pre></div>
</details>
</div>
<details class="info">
<summary>Interpreting the flagstat output</summary>
<p>This is not exactly easy to interpret. The documentation is <a href="http://www.htslib.org/doc/samtools-flagstat.html">here</a> and requires you to look through the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf"><code>sam</code> file specification</a> to understand the fields.</p>
</details>
<p>In brief, the pairs of numbers on each line represent the number of mappings with passed/failed the alignment quality conditions. This is not the same thing as the fastq quality score - when we map reads there is a scoring system used to assess how well a read sticks to the reference at a given position. Depending on how mapping is performed there may be reads which do align to a region, but too poorly for <code>samtools</code> to consider them a likely match.</p>
<p>Going through the lines of the output, the most important parts are:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Term</th>
<th style="text-align: left;">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code>X + Y in total (QC-passed reads + QC-failed reads)</code></td>
<td style="text-align: left;">Total mappings (sum of primary, secondary, and supplementary)</td>
</tr>
<tr>
<td style="text-align: left;"><code>X + Y primary</code></td>
<td style="text-align: left;">Number of reads mapped without multiple or chimeric aligments (i.e. the most reliable ones)</td>
</tr>
<tr>
<td style="text-align: left;"><code>X + Y secondary</code></td>
<td style="text-align: left;">Number of reads mapped with multiple mappings</td>
</tr>
<tr>
<td style="text-align: left;"><code>X + Y supplementary</code></td>
<td style="text-align: left;">Number of reads mapped with a chimeric alignment</td>
</tr>
<tr>
<td style="text-align: left;"><code>X + Y duplicates</code></td>
<td style="text-align: left;">Number of duplicate mappings</td>
</tr>
<tr>
<td style="text-align: left;"><code>X + Y properly paired (Z% : N/A)</code></td>
<td style="text-align: left;">Number of instances where paired reads are mapped together in the correct orientation</td>
</tr>
<tr>
<td style="text-align: left;"><code>X + Y singletons (Z% : N/A)</code></td>
<td style="text-align: left;">Number of instances of a read mapping when its paired partner failed to do so</td>
</tr>
</tbody>
</table>
<div class="admonition info">
<p class="admonition-title">Understanding the 'duplicates' term</p>
<p>The term duplicates is used to represent instances where two or more <em>identical</em> sequences mapped to the region, not different reads mapping to the same region in a manner that overlaps. In these instances, it is phenominally rare to see two identical reads produced as part of library preparation and they are much more commonly due to PCR artefacts which occur during library preparation, or some form of read duplication during sequencing.</p>
<p>In some instances it is worth identifying and removing these duplication events using a tool like <code>picard</code> (<a href="https://broadinstitute.github.io/picard/">documentation</a>).</p>
<p>If you are working with amplicon sequence, where you have produced a PCR product which is then sent for sequencing your duplication rate will be high but this is an expected consequence of the method used to produce your input DNA. In such cases a high duplication rate is expected.</p>
</div>
<hr>
<h3 id="producing-a-coverage-table-from-a-complete-alignment">Producing a coverage table from a complete alignment<a class="headerlink" href="#producing-a-coverage-table-from-a-complete-alignment" title="Permanent link">¶</a></h3>
<p>Now that we have seen a high-level overview of the mapping in the table, how do we get something actionable out of this? One of the common tasks we need to perform with a mapping file is to assess the depth of coverage over the reference sequence to determine</p>
<div class="admonition question">
<p class="admonition-title">Is there sufficient depth of coverage to be confident in the consensus sequence of our reads?</p>
</div>
<div class="admonition question">
<p class="admonition-title">Is the mapping depth even across the reference sequence, or does it pile up in hotspots?</p>
<p>Piling into hotspots can be indicative of poor mapping, where a low complexity region of a sequence (like the poly-A tail on a virus) is attracting a lot of non-specific binding.</p>
<p>In a case like this you would not want to infer the average coverage by dividing the number of mapped reads by the reference length as there are likely to be spots with no mapping.</p>
</div>
<p>One of the simplest ways to view the per-position coverage of a mapping result is to use the <code>samtools depth</code> command, which reads through the <code>bam</code> file and reports the number of nucleotides mapped at each position of the reference sequence. It is very quick to run, and the output is a simple text table which can be imported into <code>Excel</code> or <code>R</code> to calculate statistics.</p>
<p>By default the tool prints to the command line which is not ideal, so we will capture the results with a redirection.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>samtools<span class="w"> </span>depth<span class="w"> </span>-a<span class="w"> </span>SRR18260232.clean_reads.bam<span class="w"> </span>&gt;<span class="w"> </span>SRR18260232.clean_reads.txt
</code></pre></div>
</div>
<p>Running the <code>head</code> command should print the first 10 lines of the file to the command line. This is enough to see the structure of the file.</p>
<div class="admonition file-code">
<p class="admonition-title">First 10 lines of <code>SRR18260232.clean_reads.txt</code></p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>NZ_LAUS01000004.1       1       6
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>NZ_LAUS01000004.1       2       9
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>NZ_LAUS01000004.1       3       9
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>NZ_LAUS01000004.1       4       9
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>NZ_LAUS01000004.1       5       9
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>NZ_LAUS01000004.1       6       9
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>NZ_LAUS01000004.1       7       9
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>NZ_LAUS01000004.1       8       9
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a>NZ_LAUS01000004.1       9       9
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a>NZ_LAUS01000004.1       10      9
</code></pre></div>
</div>
<p>Although there are no column headers, there are only three columns in the file so it is easy to just describe the contents. The first column is the name of the sequence in the reference file which is being reported. If you have a single sequence in your reference then this will be a single value throughout the file. If you are mapping to a draft reference genome, or a genome with plasmids or other non-chromosomal elements, the content of this column will change as you move through the file.</p>
<p>The second column is the nucleotide position in the reference sequence, and the third column is the number of nucleotides mapped against that position. From the command line it is hard to see the maximum depth, but we can use a quick combination of command line tools to tell us something of the mapping:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>cut<span class="w"> </span>-f3<span class="w"> </span>SRR18260232.clean_reads.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n1
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>27988 0
</code></pre></div>
</details>
</div>
<p>What this command does is take the third column of the table (<code>cut</code>), sort the values (<code>sort</code>), and then tally up the unique instances and count them (<code>uniq -c</code>).</p>
<p>Which means that there are 27,988 positions in the reference with 0 nucleotides mapped. This tells us that there are some gaps in our mapping. Given that the reference sequence is 254,357 nucleotides long this means that about 11% of our genome does not have reads that correspond to it.</p>
<p>If you wanted to push this further, you could import the summary table into a tool like <code>Excel</code> or <code>R</code> and plot the depth of coverage along the genome or calculate the average depth and variance to decide if you wanted to trust this alignment.</p>
<p>We will not be pursuring this further in this workshop, as plotting is a fairly easy task which you can perform by yourself if you are interested.</p>
<hr>
<h3 id="producing-a-coverage-table-from-a-specific-region-of-an-alignment">Producing a coverage table from a specific region of an alignment<a class="headerlink" href="#producing-a-coverage-table-from-a-specific-region-of-an-alignment" title="Permanent link">¶</a></h3>
<p>Viewing the depth across the full genome is helpful if you're working with a small genome, such as a virus, but for larger organisms it's probably not going to be very informative. Depending on your application, you may not even want to view the coverage over an entire genome - there may be a specific marker gene or region of interest you wish to examine and as long as you have good coverage in this area, then that is sufficient for your application.</p>
<p>It is only a minor step to modify the previous <code>samtools</code> command to restrict the mapping down to a single region, or set of regions. We can do this by modifying the <code>samtools depth</code> command in one of two ways:</p>
<ol>
<li>Adding the region of interest to be reported to the command line</li>
<li>Specifying a <code>bed</code> file that contains the region of interest.</li>
</ol>
<p>The second option is good if it is a region we are going to be examining often, such as a known marker for a BAU test. It is also useful when specifying multiple regions, since it is essentially a small text table of start/stop coordinates. For simplicity we are not going to pursue this step in this training but if this is something you wish to do in your own work, you can find the documentation for writing <code>bed</code> files online <a href="https://asia.ensembl.org/info/website/upload/bed.html">here</a>.</p>
<p>For now, we are going to extract the depth information over two genes of the reference genome. There is no real trick to finding these gene coordinates - you simple need to read through an appropriate annotation file (such as the NCBI <code>genbank</code> file for the reference genome) and find them. This can also be done by downloading and viewing the genome in <code>Geneious</code>.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Gene</th>
<th style="text-align: left;">Sequence</th>
<th style="text-align: center;">Start position</th>
<th style="text-align: center;">End position</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Translation elongation factor 4</td>
<td style="text-align: left;">NZ_LAUS01000004.1</td>
<td style="text-align: center;">94,986</td>
<td style="text-align: center;">96,779</td>
</tr>
<tr>
<td style="text-align: left;">16S ribosomal RNA</td>
<td style="text-align: left;">NZ_LAUS01000004.1</td>
<td style="text-align: center;">41,621</td>
<td style="text-align: center;">43,134</td>
</tr>
</tbody>
</table>
<p>The syntax for performing this kind of depth summary is to add a parameter specifying the name of the sequence carrying the region, followed by the start and stop coordinates to be reported in the form <code>SEQUENCE:START-STOP</code>. We must also <em>index</em> the <code>bam</code> file, so that <code>samtools</code> is able to quickly parse the file, skipping the regions we are not interested in. Indexing will produce a file with the extension <code>bai</code>. We do not need to hold onto this file for long, it is quick to generate and not very large on disk space.</p>
<div class="admonition info">
<p>In our case there is only a single sequence in the reference genome, so all mapping must be to this lone sequence. Nevertheless, we must still specify the sequence as part of the command</p>
</div>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>samtools<span class="w"> </span>index<span class="w"> </span>SRR18260232.clean_reads.bam
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>samtools<span class="w"> </span>depth<span class="w"> </span>-a<span class="w"> </span>-r<span class="w"> </span>NZ_LAUS01000004.1:94986-96779<span class="w"> </span>SRR18260232.clean_reads.bam<span class="w">  </span>&gt;<span class="w"> </span>SRR18260232.clean_reads.tef4.txt
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>wc<span class="w"> </span>-l<span class="w"> </span>SRR18260232.clean_reads.txt<span class="w"> </span>SRR18260232.clean_reads.tef4.txt
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>254357 SRR18260232.clean_reads.txt
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>1794 SRR18260232.clean_reads.tef4.txt
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>256151 total
</code></pre></div>
</details>
</div>
<p>As you can see, this has cut the mapping space down considerably. It is now of a size that, if this was your marker of interest, you could easily plot this information and see the mapping depth along the TEF4 gene:</p>
<details class="info">
<summary>Plotting code (interest only)</summary>
<p>Here's a short snippet of <code>python</code> code to create a plot of the depth along the positions in the <strong>tef4</strong> file.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">plotly.express</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">px</span>
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"SRR18260232.clean_reads.tef4.txt"</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">"</span><span class="se">\t</span><span class="s2">"</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">"Chr"</span><span class="p">,</span> <span class="s2">"Position"</span><span class="p">,</span> <span class="s2">"Depth"</span><span class="p">])</span>
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>
<a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a><span class="n">fig</span> <span class="o">=</span> <span class="n">px</span><span class="o">.</span><span class="n">area</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">"Position"</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">"Depth"</span><span class="p">)</span>
<a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a><span class="n">fig</span><span class="o">.</span><span class="n">write_image</span><span class="p">(</span><span class="s2">"SRR18260232.clean_reads.tef4.svg"</span><span class="p">)</span>
</code></pre></div>
<details class="success">
<summary>Output</summary>
<p></p><center><a class="glightbox" data-type="image" data-width="auto" data-height="auto" href="../../img/level2_35_tef4.svg" data-desc-position="bottom"><img alt="SRR18260232 TEF4 depth across gene" src="../../img/level2_35_tef4.svg"></a></center><p></p>
</details>
</div>
</details>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>Create a 16S rRNA sequence coverage report using the <code>SRR18260232.raw_reads.bam</code> mapping file as your starting point.</p>
<p>If you're interested, also produce a small plot of the results.</p>
<details class="circle-check">
<summary>Solution</summary>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>samtools<span class="w"> </span>index<span class="w"> </span>SRR18260232.raw_reads.bam
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>samtools<span class="w"> </span>depth<span class="w"> </span>-a<span class="w"> </span>-r<span class="w"> </span>NZ_LAUS01000004.1:41621-43134<span class="w"> </span>SRR18260232.raw_reads.bam<span class="w">  </span>&gt;<span class="w"> </span>SRR18260232.raw_reads.16s_rrna.txt
</code></pre></div>
</div>
</details>
</div>
<hr>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../34_mapping_filters/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 3.4 - Filtering and compressing sam files">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                3.4 - Filtering and compressing sam files
              </div>
            </div>
          </a>
        
        
          
          <a href="../41_annotation_overview/" class="md-footer__link md-footer__link--next" aria-label="Next: 4.1 - Prediction of protein coding sequences">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                4.1 - Prediction of protein coding sequences
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Intermediate Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate", "navigation.tabs", "navigation.tabs.sticky"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>