<!DOCTYPE html><html lang="en" class="no-js"><head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="MPI, Genomics Aotearoa &amp; NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/hts_workshop_mpi/level2/45_annotation_kraken2/">
      
      
        <link rel="prev" href="../44_annotation_protein/">
      
      
        <link rel="next" href="../../proficiency/2023_proficiency/">
      
      
      <link rel="icon" href="../../theme_images/icon_fig.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.21">
    
    
      
        <title>4.5 - Classification of sequences with k-mer composition profiles - High Throughput Sequencing</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2a3383ac.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&amp;display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"><script src="../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#45-classification-of-sequences-with-k-mer-composition-profiles" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="High Throughput Sequencing" class="md-header__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            High Throughput Sequencing
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              4.5 - Classification of sequences with k-mer composition profiles
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to light mode" type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to dark mode" type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to system preference" type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../level1/11_nesi_connection/" class="md-tabs__link">
          
  
  
  Level1

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../11_shell_manipulation/" class="md-tabs__link">
          
  
  
  Level2

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../proficiency/2023_proficiency/" class="md-tabs__link">
          
  
  
  Proficiency

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../supplementary/common_terms/" class="md-tabs__link">
          
  
  
  Supplementary

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="High Throughput Sequencing" class="md-nav__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    High Throughput Sequencing
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2">
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Level1
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Level1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/11_nesi_connection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Connecting to NeSI
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/12_shell_navigation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Introducing the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/13_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/21_illumina_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Assesing Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/22_illumina_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Filtering Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/31_nanopore_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Assesing Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/32_nanopore_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/41_blastn_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Annotating with BLAST
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/42_slurm_introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Introduction to slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level1/43_blast_interpretation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Level2
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Level2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../11_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../12_shell_variables/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Working with variables and loops
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../13_shell_redirection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Working with redirection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../14_nesi_environment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../21_assembly_de_novo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Genome Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../22_assembly_spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Assembling short-read Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../23_assembly_flye/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Assembling Oxford Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../24_assembly_evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Evaluating the results of an assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../25_assembly_polishing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.5 - Polishing of Oxford Nanopore assemblies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../31_coverage_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.1 - Overview of mapping sequences to a reference genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../32_illumina_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Mapping Illumina sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../33_nanopore_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.3 - Mapping Oxford Nanopore sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../34_mapping_filters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.4 - Filtering and compressing sam files
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../35_mapping_statistics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../41_annotation_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Prediction of protein coding sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../42_annotation_prodigal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Prediction with prodigal
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../43_annotation_augustus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Prediction with AUGUSTUS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../44_annotation_protein/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#why-not-blast" class="md-nav__link">
    <span class="md-ellipsis">
      Why not BLAST?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#getting-started-with-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Getting started with kraken2
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#performing-basic-classification-with-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Performing basic classification with kraken2
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-kraken2-output-format" class="md-nav__link">
    <span class="md-ellipsis">
      The kraken2 output format
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#refining-the-kraken2-classification" class="md-nav__link">
    <span class="md-ellipsis">
      Refining the kraken2 classification
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#comparing-the-outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Comparing the outputs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-reliable-are-these-results" class="md-nav__link">
    <span class="md-ellipsis">
      How reliable are these results?
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4">
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proficiency
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Proficiency
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 1)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency_alternate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 2)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5">
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Supplementary
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/common_terms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Common terms
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/fastq_format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The FASTQ file format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/slurm_module_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm module guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#why-not-blast" class="md-nav__link">
    <span class="md-ellipsis">
      Why not BLAST?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#getting-started-with-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Getting started with kraken2
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#performing-basic-classification-with-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Performing basic classification with kraken2
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-kraken2-output-format" class="md-nav__link">
    <span class="md-ellipsis">
      The kraken2 output format
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#refining-the-kraken2-classification" class="md-nav__link">
    <span class="md-ellipsis">
      Refining the kraken2 classification
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#comparing-the-outputs" class="md-nav__link">
    <span class="md-ellipsis">
      Comparing the outputs
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-reliable-are-these-results" class="md-nav__link">
    <span class="md-ellipsis">
      How reliable are these results?
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="45-classification-of-sequences-with-k-mer-composition-profiles">4.5 - Classification of sequences with <em>k</em>-mer composition profiles<a class="headerlink" href="#45-classification-of-sequences-with-k-mer-composition-profiles" title="Permanent link">¶</a></h1>
<div class="admonition clock">
<p class="admonition-title">time</p>
<ul>
<li>Teaching: 15 minutes</li>
<li>Teaching: 15 minutes</li>
</ul>
</div>
<div class="admonition circle-info">
<p class="admonition-title">Objectives and Key points</p>
<h4 id="objectives">Objectives<a class="headerlink" href="#objectives" title="Permanent link">¶</a></h4>
<ul>
<li>Understand when we may need to use something other than <code>BLAST</code> for getting quick classification results.</li>
<li>Know how to run <code>kraken2</code> using a standard database, and interpret the output files.</li>
</ul>
<h4 id="keypoints">Keypoints<a class="headerlink" href="#keypoints" title="Permanent link">¶</a></h4>
<ul>
<li><em>k</em>-mer based classification tools provide a rapid means of assigning taxonomy to sequence data.</li>
<li>The quality of the classification is tool and database dependent, results are also harder to interpret than a BLAST output</li>
<li>For example, there are no identity/coverage statistics from which judgement calls can be made.</li>
</ul>
</div>
<hr>
<h2 id="why-not-blast">Why not <code>BLAST</code>?<a class="headerlink" href="#why-not-blast" title="Permanent link">¶</a></h2>
<p>As we saw in the level 1 training, running <code>BLASTn</code> to classify sequences is a good way to get annotation information about a sequence, and from this we can infer taxonomic origin of the sequence. However, obtaining classification information is not the primary purpose of <code>BLASTn</code> so we need to perform some manual inspection of the output to make a call on the classification of each sequence.</p>
<p>When we do not care about functional information, and <em>only</em> want classification of sequences, there are a suite of tools designed specifically with this purpose in mind. Rather than perform pairwise sequence alignment between query and target sequences, these tools process their reference database into a compressed form containing sequence fingerprints.</p>
<p>Query sequences are fingerprinted and the results compared to the database providing the most likely, <strong><em>lowest common ancestor</em></strong> for each query sequence.</p>
<div class="admonition note">
<p class="admonition-title">Options for <em>k</em>-mer classification</p>
<p>How this fingerprinting is done is tool specific, and goes outside the realm of biology. There are a number of tools developed for this kind of work, including;</p>
<ol>
<li><a href="https://github.com/DerrickWood/kraken2/">kraken2</a> (<a href="https://doi.org/10.1186/s13059-019-1891-0">Wood <em>et al</em>., 2019</a>)</li>
<li><a href="http://clark.cs.ucr.edu/">CLARK</a> (<a href="https://doi.org/10.1186/s12864-015-1419-2">Ounit <em>et al</em>., 2015</a>)</li>
<li><a href="https://github.com/muellan/metacache">MetaCache</a> (<a href="https://doi.org/10.1093/bioinformatics/btx520">Müller <em>et al</em>., 2017</a>)</li>
<li><a href="https://gitlab.com/rki_bioinformatics/ganon">ganon</a> (<a href="https://doi.org/10.1093/bioinformatics/btaa458">Piro <em>et al</em>., 2020</a>)</li>
</ol>
</div>
<p>We will be using <code>kraken2</code> today, as it is a tool which is used often within our operations.</p>
<hr>
<h2 id="getting-started-with-kraken2">Getting started with <code>kraken2</code><a class="headerlink" href="#getting-started-with-kraken2" title="Permanent link">¶</a></h2>
<p>Navigate to <code>/nesi/project/nesi03181/phel/USERNAME/level2/annotation_kraken2/</code> and prepare to run <code>kraken2</code>.</p>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>Find and load the most current version of <code>kraken2</code>. Once loaded, run with the help paramter (<code>-h</code>) to confirm that the module has successfully loaded.</p>
<details class="circle-check">
<summary>Solution</summary>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>module<span class="w"> </span>spider<span class="w"> </span>kraken2
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>module<span class="w"> </span>purge
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>module<span class="w"> </span>load<span class="w"> </span>Kraken2/2.1.3-GCC-11.3.0
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>kraken2<span class="w"> </span>-h
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>Usage: kraken2 [options] &lt;filename(s)&gt;
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>Options:
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>--db NAME               Name for Kraken 2 DB
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>                        (default: "/opt/nesi/db/Kraken2/standard-2018-09")
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>--threads NUM           Number of threads (default: 1)
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>--quick                 Quick operation (use first hit or hits)
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>--unclassified-out FILENAME
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>                        Print unclassified sequences to filename
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>--classified-out FILENAME
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>                        Print classified sequences to filename
<a id="__codelineno-1-12" name="__codelineno-1-12" href="#__codelineno-1-12"></a>--output FILENAME       Print output to filename (default: stdout); "-" will
<a id="__codelineno-1-13" name="__codelineno-1-13" href="#__codelineno-1-13"></a>                        suppress normal output
<a id="__codelineno-1-14" name="__codelineno-1-14" href="#__codelineno-1-14"></a>--confidence FLOAT      Confidence score threshold (default: 0.0); must be
<a id="__codelineno-1-15" name="__codelineno-1-15" href="#__codelineno-1-15"></a>                        in [0, 1].
<a id="__codelineno-1-16" name="__codelineno-1-16" href="#__codelineno-1-16"></a>--minimum-base-quality NUM
<a id="__codelineno-1-17" name="__codelineno-1-17" href="#__codelineno-1-17"></a>                        Minimum base quality used in classification (def: 0,
<a id="__codelineno-1-18" name="__codelineno-1-18" href="#__codelineno-1-18"></a>                        only effective with FASTQ input).
<a id="__codelineno-1-19" name="__codelineno-1-19" href="#__codelineno-1-19"></a>--report FILENAME       Print a report with aggregrate counts/clade to file
<a id="__codelineno-1-20" name="__codelineno-1-20" href="#__codelineno-1-20"></a>--use-mpa-style         With --report, format report output like Kraken 1's
<a id="__codelineno-1-21" name="__codelineno-1-21" href="#__codelineno-1-21"></a>                        kraken-mpa-report
<a id="__codelineno-1-22" name="__codelineno-1-22" href="#__codelineno-1-22"></a>--report-zero-counts    With --report, report counts for ALL taxa, even if
<a id="__codelineno-1-23" name="__codelineno-1-23" href="#__codelineno-1-23"></a>                        counts are zero
<a id="__codelineno-1-24" name="__codelineno-1-24" href="#__codelineno-1-24"></a>--report-minimizer-data With --report, report minimizer and distinct minimizer
<a id="__codelineno-1-25" name="__codelineno-1-25" href="#__codelineno-1-25"></a>                        count information in addition to normal Kraken report
<a id="__codelineno-1-26" name="__codelineno-1-26" href="#__codelineno-1-26"></a>--memory-mapping        Avoids loading database into RAM
<a id="__codelineno-1-27" name="__codelineno-1-27" href="#__codelineno-1-27"></a>--paired                The filenames provided have paired-end reads
<a id="__codelineno-1-28" name="__codelineno-1-28" href="#__codelineno-1-28"></a>--use-names             Print scientific names instead of just taxids
<a id="__codelineno-1-29" name="__codelineno-1-29" href="#__codelineno-1-29"></a>--gzip-compressed       Input files are compressed with gzip
<a id="__codelineno-1-30" name="__codelineno-1-30" href="#__codelineno-1-30"></a>--bzip2-compressed      Input files are compressed with bzip2
<a id="__codelineno-1-31" name="__codelineno-1-31" href="#__codelineno-1-31"></a>--minimum-hit-groups NUM
<a id="__codelineno-1-32" name="__codelineno-1-32" href="#__codelineno-1-32"></a>                        Minimum number of hit groups (overlapping k-mers
<a id="__codelineno-1-33" name="__codelineno-1-33" href="#__codelineno-1-33"></a>                        sharing the same minimizer) needed to make a call
<a id="__codelineno-1-34" name="__codelineno-1-34" href="#__codelineno-1-34"></a>                        (default: 2)
<a id="__codelineno-1-35" name="__codelineno-1-35" href="#__codelineno-1-35"></a>--help                  Print this message
<a id="__codelineno-1-36" name="__codelineno-1-36" href="#__codelineno-1-36"></a>--version               Print version information
<a id="__codelineno-1-37" name="__codelineno-1-37" href="#__codelineno-1-37"></a>
<a id="__codelineno-1-38" name="__codelineno-1-38" href="#__codelineno-1-38"></a>If none of the *-compressed flags are specified, and the filename provided
<a id="__codelineno-1-39" name="__codelineno-1-39" href="#__codelineno-1-39"></a>is a regular file, automatic format detection is attempted.
</code></pre></div>
</details>
</div>
</details>
</div>
<p>One of the nice perks of working with <code>kraken2</code> is that the maintainers of the tool provide pre-computed, up-to-date <a href="https://benlangmead.github.io/aws-indexes/k2">databases for classification</a>. These are released in many different variants - some for use of HPCs and some for local machines, and databases which cover a range of target organisms.</p>
<div class="admonition info">
<p class="admonition-title">We use <code>PlusPFP</code> database, which contains the following lineages:</p>
<ol>
<li>Archaea</li>
<li>Bacteria</li>
<li>Fungi</li>
<li>Human</li>
<li>Plant</li>
<li>Protozoa</li>
<li>Virus</li>
<li>Plasmid sequences</li>
<li>Sequencing and transformation vectors (UniVec)</li>
</ol>
<hr>
<p>For training, we are using a reduced version of this database (<code>PlusPFP-16</code>) so that our jobs do not need to queue as long, but the same lineages are represented in this smaller database.</p>
</div>
<p>This makes it a comprehensive database for most purposes, but if you are tring to annotate insect sequences this is not the right database for you. For training purposes, a copy of this database can be found at <code>/nesi/project/nesi03181/phel/databases/</code> and we also have a copy in our diagnostic workspace.</p>
<div class="admonition warning">
<p class="admonition-title">Make sure you have the right database for your work</p>
<p>It is critical to know what is in your classification database, as a failure to include the correct reference material in the database will result in sequences being either misclassified, or not classified at all.</p>
</div>
<hr>
<h2 id="performing-basic-classification-with-kraken2">Performing basic classification with <code>kraken2</code><a class="headerlink" href="#performing-basic-classification-with-kraken2" title="Permanent link">¶</a></h2>
<p>To begin, we are going to write a <code>slurm</code> script to run <code>kraken2</code> with a fairly minimal set in input parameters just to see the tool in action. Although <code>kraken2</code> is very fast to run, it requires a lot of memory (RAM) to run so we cannot run it from the command line in most cases.</p>
<p>Complete the following script, then submit your <code>slurm</code> job.</p>
<div class="admonition file-code">
<p class="admonition-title">kraken2_basic.sl</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="c1">#SBATCH --account       nesi03181</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="c1">#SBATCH --job-name      kraken2_basic</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="c1">#SBATCH --time          00:01:00</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="c1">#SBATCH --cpus-per-task 16</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="c1">#SBATCH --mem           20G</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a><span class="c1">#SBATCH --error         kraken2_basic.%j.err</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a><span class="c1">#SBATCH --output        kraken2_basic.%j.out</span>
<a id="__codelineno-2-9" name="__codelineno-2-9" href="#__codelineno-2-9"></a>
<a id="__codelineno-2-10" name="__codelineno-2-10" href="#__codelineno-2-10"></a>module<span class="w"> </span>purge
<a id="__codelineno-2-11" name="__codelineno-2-11" href="#__codelineno-2-11"></a>module<span class="w"> </span>load<span class="w"> </span>Kraken2/2.1.3-GCC-11.3.0
<a id="__codelineno-2-12" name="__codelineno-2-12" href="#__codelineno-2-12"></a>
<a id="__codelineno-2-13" name="__codelineno-2-13" href="#__codelineno-2-13"></a><span class="nb">cd</span><span class="w"> </span>/nesi/project/nesi03181/phel/USERNAME/level2/annotation_kraken2/
<a id="__codelineno-2-14" name="__codelineno-2-14" href="#__codelineno-2-14"></a>
<a id="__codelineno-2-15" name="__codelineno-2-15" href="#__codelineno-2-15"></a><span class="nv">PFP_DB</span><span class="o">=</span><span class="s2">"/nesi/project/nesi03181/phel/databases/k2_pluspfp_16gb_20231009"</span>
<a id="__codelineno-2-16" name="__codelineno-2-16" href="#__codelineno-2-16"></a>kraken2<span class="w"> </span>--db<span class="w"> </span><span class="si">${</span><span class="nv">PFP_DB</span><span class="si">}</span><span class="w"> </span>--threads<span class="w"> </span><span class="si">${</span><span class="nv">SLURM_CPUS_PER_TASK</span><span class="si">}</span><span class="w"> </span>--use-names<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-17" name="__codelineno-2-17" href="#__codelineno-2-17"></a><span class="w">    </span>--output<span class="w"> </span>outputs/input_seqs.out<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-18" name="__codelineno-2-18" href="#__codelineno-2-18"></a><span class="w">    </span>input/input_seqs.fna
</code></pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">No email block in the <code>slurm</code> script</p>
<p>This job will take less than 30 seconds to complete, so we're not going to bother with the email notifications for job status. The only reason we need to push this job through <code>slurm</code> at all is because of the memory required to run the tool.</p>
</div>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>sbatch<span class="w"> </span>kraken2_basic.sl
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>Submitted batch job XXXXXXXX
</code></pre></div>
</details>
</div>
<p>This will not take long to run.</p>
<hr>
<h2 id="the-kraken2-output-format">The <code>kraken2</code> output format<a class="headerlink" href="#the-kraken2-output-format" title="Permanent link">¶</a></h2>
<p>Once the job is complete, use <code>less</code> or <code>head</code> to take a look at the output file. The output file is a tab-delimited table with the following columns:</p>
<div class="admonition file-code">
<p class="admonition-title">kraken2_basic.sl</p>
<table>
<thead>
<tr>
<th style="text-align: center;">Column</th>
<th style="text-align: left;">Content</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">1</td>
<td style="text-align: left;">A single letter (C or U) designating whether the sequence was classified or not</td>
</tr>
<tr>
<td style="text-align: center;">2</td>
<td style="text-align: left;">Sequence name</td>
</tr>
<tr>
<td style="text-align: center;">3</td>
<td style="text-align: left;">The <a href="https://www.ncbi.nlm.nih.gov/taxonomy">NCBI taxonomy ID</a> applied to the sequence (0 if unclassified)<br>Because we used the <code>--use-names</code> flag, this takes the form <code>TAXONOMY NAME (TAXONOMY ID)</code><br>By default this is just the numeric value</td>
</tr>
<tr>
<td style="text-align: center;">4</td>
<td style="text-align: left;">Sequence length<br>For paired-end data, this will be in the form <code>FORWARD|REVERSE</code></td>
</tr>
<tr>
<td style="text-align: center;">5</td>
<td style="text-align: left;">A list of how many <em>k</em>-mers mapped to each taxonomy ID<br>Results are reported as colon-delimited pairings of <code>TAXID:NUMBER</code></td>
</tr>
</tbody>
</table>
</div>
<p>It is also possible to generate a report file that summarises the number of sequences classified to each taxonomy/rank, using the modified command:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>kraken2<span class="w"> </span>--db<span class="w"> </span><span class="si">${</span><span class="nv">PFP_DB</span><span class="si">}</span><span class="w"> </span>--threads<span class="w"> </span><span class="si">${</span><span class="nv">SLURM_CPUS_PER_TASK</span><span class="si">}</span><span class="w"> </span>--use-names<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="w">    </span>--output<span class="w"> </span>outputs/input_seqs.out<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="w">    </span>--report<span class="w"> </span>outputs/input_seqs.report.txt<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="w">    </span>input/input_seqs.fna
</code></pre></div>
</div>
<p>Depending on your situation, this may be more useful.</p>
<div class="admonition info">
<p class="admonition-title">Classification output versus report</p>
<p><strong>When would you need the per-sequence output</strong></p>
<p>If you are screening a sequence set for the specific reads obtained from a particular organism, or are trying to remove contamination from a sample, having the per-sequence results are invaluable.</p>
<p>The content of this file can be easily parsed with the <code>cut</code> and <code>grep</code> tools to extract lists of sequences match or failing to match different search criteria.</p>
<hr>
<p><strong>When would you need the high-level report</strong></p>
<p>If you trust that your data is uncontaminated, you might be trying ot achive one of the following:</p>
<ol>
<li>Detecting the presence of a particular organism in a mixed population (i.e. eDNA surveillance).</li>
<li>Confirming the identity of a pure culture.</li>
<li>In this case you might be comfortable taking a majority-rules approach to your data - if most of the sequences are classified as a single species, this is likely the identity of your culture.</li>
</ol>
</div>
<hr>
<h2 id="refining-the-kraken2-classification">Refining the <code>kraken2</code> classification<a class="headerlink" href="#refining-the-kraken2-classification" title="Permanent link">¶</a></h2>
<p>By default <code>kraken2</code> does not apply any filtering to the classification data. Reads are classified based on the most commonly reported taxonomy for a sequence, with no regard for how good the evidence for this classification can be. There is a built-in parameter to increase the stringency of classification. The filter is applied during classification and cannot be run after the fact.</p>
<p>This means that if you want to change your filtering criteria you must perform classification again. This differs from tools like <code>BLAST</code>, where a more stringent filter can be applied after running classification.</p>
<blockquote>
<p>For example, if we run a <code>BLAST</code> search looking to report all hits with greater than 30% identity, we can later filter the table to keep results with greater than 50% identity by filtering the appropriate column.</p>
</blockquote>
<div class="admonition warning">
<p class="admonition-title">Limitations with the <code>kraken2</code> filtering</p>
<p>It is also difficult to interpret the numeric values of filters.</p>
<p>According to the <code>kraken2</code> manual, the confidence scoring does not have a simple interpretation.</p>
<p>For example, with a <code>BLAST</code> identitiy of 90%, we can infer that 90% of positions are identical between query and target match, or a bootstrap support of 90% would mean that 90% of the resamplings agree with the full result.</p>
<p>However, there is no such interpretation of what a <code>0.9</code> confidence value in <code>kraken2</code> means, other than it is more strict that a value of <code>0.5</code>.</p>
</div>
<p>We generally don't need to apply much of a cutoff to <code>kraken2</code> - the tool is already good at identifying sequences which can't be reliably classified.</p>
<div class="admonition file-code">
<p class="admonition-title">kraken2_refined.sl</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="ch">#!/bin/bash -e</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="c1">#SBATCH --account       nesi03181</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="c1">#SBATCH --job-name      kraken2_refined</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="c1">#SBATCH --time          00:01:00</span>
<a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a><span class="c1">#SBATCH --cpus-per-task 16</span>
<a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a><span class="c1">#SBATCH --mem           20G</span>
<a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a><span class="c1">#SBATCH --error         kraken2_refined.%j.err</span>
<a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a><span class="c1">#SBATCH --output        kraken2_refined.%j.out</span>
<a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a>
<a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a>module<span class="w"> </span>purge
<a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a>module<span class="w"> </span>load<span class="w"> </span>Kraken2/2.1.3-GCC-11.3.0
<a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a>
<a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a><span class="nb">cd</span><span class="w"> </span>/nesi/project/nesi03181/phel/USERNAME/level2/annotation_kraken2/
<a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a>
<a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a><span class="nv">PFP_DB</span><span class="o">=</span><span class="s2">"/nesi/project/nesi03181/phel/databases/k2_pluspfp_16gb_20231009"</span>
<a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a>kraken2<span class="w"> </span>--db<span class="w"> </span><span class="si">${</span><span class="nv">PFP_DB</span><span class="si">}</span><span class="w"> </span>--threads<span class="w"> </span><span class="si">${</span><span class="nv">SLURM_CPUS_PER_TASK</span><span class="si">}</span><span class="w"> </span>--confidence<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span>--use-names<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a><span class="w">    </span>--output<span class="w"> </span>outputs/input_seqs.conf_0.1.out<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a><span class="w">    </span>input/input_seqs.fna
</code></pre></div>
</div>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>sbatch<span class="w"> </span>kraken2_refined.sl
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>Submitted batch job XXXXXXXX
</code></pre></div>
</details>
</div>
<hr>
<h2 id="comparing-the-outputs">Comparing the outputs<a class="headerlink" href="#comparing-the-outputs" title="Permanent link">¶</a></h2>
<p>Without going into too much detail, we can briefly compare the contents of both output files. For starters, we can use some of the basic <code>bash</code> command line tools to work out how many sequences were in our input and output files.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">"&gt;"</span><span class="w"> </span>input/input_seqs.fna<span class="w"> </span>
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>5614
</code></pre></div>
</details>
</div>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>wc<span class="w"> </span>-l<span class="w"> </span>outputs/input_seqs.*<span class="w"> </span>
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a> 5614 outputs/input_seqs.conf_0.1.out
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a> 5614 outputs/input_seqs.out
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a>11228 total
</code></pre></div>
</details>
</div>
<p>The number of results in each output file is the same, and also the same as the number of sequences in the input file. This is different to <code>BLAST</code>, where an unclassified or unannotated sequence is not reported in the output.</p>
<p>A quick way to look for differences in the classification rate of the unfiltered and filtered <code>kraken2</code> output is to apply the following commands:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>cut<span class="w"> </span>-f<span class="w"> </span><span class="si">${</span><span class="nv">n</span><span class="si">}</span><span class="w"> </span>outputs/input_seqs.out<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
</div>
<p>If we point this towards the correct column, we can count how many sequences were successfully classified or not.</p>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>Look through these tutorial notes, or the <code>kraken2</code> documentation, to find the appropriate column index for the command above. Once you have determined the correct value for <code>${n}</code>, complete the command above and run it for both <code>kraken2</code> outputs.</p>
<p>How many sequences were classified for each run?</p>
<details class="circle-check">
<summary>Solution</summary>
<p>The first column (C or U values) is the easiest one to work with, although you could also get the same result from column 3.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="nv">n</span><span class="o">=</span><span class="m">1</span>
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a>cut<span class="w"> </span>-f<span class="w"> </span><span class="si">${</span><span class="nv">n</span><span class="si">}</span><span class="w"> </span>outputs/input_seqs.out<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>cut<span class="w"> </span>-f<span class="w"> </span><span class="si">${</span><span class="nv">n</span><span class="si">}</span><span class="w"> </span>outputs/input_seqs.conf_0.1.out<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a># outputs/input_seqs.out
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>5205 C
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a> 409 U
<a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a>
<a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a># outputs/input_seqs.conf_0.1.out
<a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a>  10 C
<a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a>5604 U
</code></pre></div>
</details>
</details>
</div>
<p>As you can see, even a small degree of filtering has a drastic impact on the classification rate. If you are applying a filter value, using a low threshold is recommended to avoid discarding too much usable data.</p>
<hr>
<h2 id="how-reliable-are-these-results">How reliable are these results?<a class="headerlink" href="#how-reliable-are-these-results" title="Permanent link">¶</a></h2>
<p>The input sequences in this exercise are simulated Oxford Nanopore reads from the following organisms;</p>
<ol>
<li>Tomato (<em>Solanum lycopersicum</em>) (5,447 sequences)</li>
<li><em>Xyllela fastidiosa</em> (166 sequences)</li>
<li>Potato virus X (1 sequence)</li>
</ol>
<p>If you skim through the results of both output files, you should note the following:</p>
<details class="file-code">
<summary>outputs/input_seqs.out</summary>
<table>
<thead>
<tr>
<th style="text-align: left;">Lineage</th>
<th style="text-align: center;">Observations</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><em>Solanum lycopersicum</em></td>
<td style="text-align: center;">4,976</td>
</tr>
<tr>
<td style="text-align: left;">unclassified</td>
<td style="text-align: center;">409</td>
</tr>
<tr>
<td style="text-align: left;"><em>Xylella fastidiosa</em></td>
<td style="text-align: center;">162</td>
</tr>
<tr>
<td style="text-align: left;"><em>Homo sapiens</em></td>
<td style="text-align: center;">37</td>
</tr>
<tr>
<td style="text-align: left;"><em>Solanum</em> subgen. <em>Lycopersicon</em></td>
<td style="text-align: center;">6</td>
</tr>
<tr>
<td style="text-align: left;"><em>Solanum pennellii</em></td>
<td style="text-align: center;">2</td>
</tr>
<tr>
<td style="text-align: left;"><em>Triticum aestivum</em></td>
<td style="text-align: center;">2</td>
</tr>
<tr>
<td style="text-align: left;"><em>Asparagus officinalis</em></td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Flavobacterium</em> sp.</td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Hordeum vulgare</em></td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Lolium rigidum</em></td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Oryza brachyantha</em></td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Pandoraea thiooxydans</em></td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;">Potato virus X</td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Setaria viridis</em></td>
<td style="text-align: center;">1</td>
</tr>
<tr>
<td style="text-align: left;"><em>Streptomyces cinnabarinus</em></td>
<td style="text-align: center;">1</td>
</tr>
</tbody>
</table>
</details>
<details class="file-code">
<summary>outputs/input_seqs.conf_0.1.out</summary>
<table>
<thead>
<tr>
<th style="text-align: left;">Lineage</th>
<th style="text-align: center;">Observations</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">unclassified</td>
<td style="text-align: center;">5,604</td>
</tr>
<tr>
<td style="text-align: left;"><em>Homo sapiens</em></td>
<td style="text-align: center;">6</td>
</tr>
<tr>
<td style="text-align: left;"><em>Solanum lycopersicum</em></td>
<td style="text-align: center;">2</td>
</tr>
<tr>
<td style="text-align: left;">Potato virus X</td>
<td style="text-align: center;">1</td>
</tr>
</tbody>
</table>
</details>
<p>What should be apparent here is that while filtering did remove a lot of the noise from our sample (we lost species we know are not truly present) we also lost a signal that <strong><em>was</em></strong> in the sample.</p>
<div class="admonition warning">
<p class="admonition-title">Recommendation for <code>kraken2</code></p>
<p>Don't let this put you off using <code>kraken2</code> - it is an excellent tool for rapidly sifting through HTS data for quick summaries. However, results should only be considered indicative in nature, not definitive.</p>
<p>Use <code>kraken2</code> to make a quick check for names of interest (regulated organisms, host, contamination) but follow up with more thorough investigation.</p>
</div>
<hr>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../44_annotation_protein/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 4.4 - Protein annotation with BLAST-like methods">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                4.4 - Protein annotation with BLAST-like methods
              </div>
            </div>
          </a>
        
        
          
          <a href="../../proficiency/2023_proficiency/" class="md-footer__link md-footer__link--next" aria-label="Next: 2023 Level 2 proficiency testing (version 1)">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                2023 Level 2 proficiency testing (version 1)
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Intermediate Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate", "navigation.tabs", "navigation.tabs.sticky"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>