<!DOCTYPE html><html lang="en" class="no-js"><head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="MPI, Genomics Aotearoa &amp; NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/hts_workshop_mpi/level1/43_blast_interpretation/">
      
      
        <link rel="prev" href="../42_slurm_introduction/">
      
      
        <link rel="next" href="../../level2/11_shell_manipulation/">
      
      
      <link rel="icon" href="../../theme_images/icon_fig.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.21">
    
    
      
        <title>4.3 - Interpreting BLAST results - High Throughput Sequencing</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2a3383ac.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&amp;display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"><script src="../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#43-interpreting-blast-results" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="High Throughput Sequencing" class="md-header__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            High Throughput Sequencing
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              4.3 - Interpreting BLAST results
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to light mode" type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to dark mode" type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to system preference" type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../11_nesi_connection/" class="md-tabs__link">
          
  
  
  Level1

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../level2/11_shell_manipulation/" class="md-tabs__link">
          
  
  
  Level2

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../proficiency/2023_proficiency/" class="md-tabs__link">
          
  
  
  Proficiency

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../supplementary/common_terms/" class="md-tabs__link">
          
  
  
  Supplementary

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="High Throughput Sequencing" class="md-nav__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    High Throughput Sequencing
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Level1
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Level1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../11_nesi_connection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Connecting to NeSI
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../12_shell_navigation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Introducing the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../13_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../21_illumina_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Assesing Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../22_illumina_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Filtering Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../31_nanopore_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Assesing Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../32_nanopore_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../41_blastn_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Annotating with BLAST
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../42_slurm_introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Introduction to slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interpretting-the-results-of-blast-queries" class="md-nav__link">
    <span class="md-ellipsis">
      Interpretting the results of BLAST queries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examining-our-file" class="md-nav__link">
    <span class="md-ellipsis">
      Examining our file
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3">
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Level2
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Level2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/11_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/12_shell_variables/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Working with variables and loops
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/13_shell_redirection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Working with redirection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/14_nesi_environment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/21_assembly_de_novo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Genome Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/22_assembly_spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Assembling short-read Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/23_assembly_flye/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Assembling Oxford Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/24_assembly_evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Evaluating the results of an assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/25_assembly_polishing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.5 - Polishing of Oxford Nanopore assemblies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/31_coverage_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.1 - Overview of mapping sequences to a reference genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/32_illumina_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Mapping Illumina sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/33_nanopore_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.3 - Mapping Oxford Nanopore sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/34_mapping_filters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.4 - Filtering and compressing sam files
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/35_mapping_statistics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/41_annotation_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Prediction of protein coding sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/42_annotation_prodigal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Prediction with prodigal
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/43_annotation_augustus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Prediction with AUGUSTUS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/44_annotation_protein/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/45_annotation_kraken2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4">
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proficiency
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Proficiency
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 1)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency_alternate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 2)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5">
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Supplementary
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/common_terms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Common terms
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/fastq_format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The FASTQ file format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/slurm_module_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm module guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#objectives" class="md-nav__link">
    <span class="md-ellipsis">
      Objectives
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#keypoints" class="md-nav__link">
    <span class="md-ellipsis">
      Keypoints
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interpretting-the-results-of-blast-queries" class="md-nav__link">
    <span class="md-ellipsis">
      Interpretting the results of BLAST queries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examining-our-file" class="md-nav__link">
    <span class="md-ellipsis">
      Examining our file
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="43-interpreting-blast-results">4.3 - Interpreting BLAST results<a class="headerlink" href="#43-interpreting-blast-results" title="Permanent link">¶</a></h1>
<div class="admonition clock">
<p class="admonition-title">time</p>
<ul>
<li>Teaching: 15 minutes</li>
<li>Exercises: 15 minutes</li>
</ul>
</div>
<div class="admonition circle-info">
<p class="admonition-title">Objectives and Key points</p>
<h4 id="objectives">Objectives<a class="headerlink" href="#objectives" title="Permanent link">¶</a></h4>
<ul>
<li>Learn how to interpret <code>BLAST</code> results </li>
<li>Understand the meaning of <em>identity</em>, <em>coverage</em>, <em>e-value</em>, and <em>bitscore</em> metrics</li>
</ul>
<h4 id="keypoints">Keypoints<a class="headerlink" href="#keypoints" title="Permanent link">¶</a></h4>
<ul>
<li>Interpreting the results of BLAST alignments requires thought and attention </li>
</ul>
</div>
<hr>
<h2 id="interpretting-the-results-of-blast-queries">Interpretting the results of BLAST queries<a class="headerlink" href="#interpretting-the-results-of-blast-queries" title="Permanent link">¶</a></h2>
<p>It is important to remember, like most bioinformatics tools, <code>BLAST</code> has a specific job. In this case sequence alignment to a set of references. <code>BLAST</code> is really good at this job, but it <strong><em>does not offer interpretation of its alignments</em></strong>.</p>
<p>Interpretation is completely up to the user on a case by case basis. It is therefore important to know your data and to understand the output metrics given by <code>BLAST</code> to help you make a biologically usefull interpretation of the results. </p>
<p>There are typically four main metrics that we need ot check when reviewing a <code>BLAST</code> result:</p>
<div class="admonition info">
<p class="admonition-title">Key <code>BLAST</code> metrics</p>
<p><strong>Coverage</strong></p>
<ul>
<li>This value tells us as a percentage how much of our query aligns with the database match. </li>
<li>A small coverage value means only a small part of our sequence has matched. A perfect match would have a coverage of 100. </li>
</ul>
<p><strong>Identity</strong></p>
<ul>
<li>This represents the percent of bases which are identical between the query and the database hit, <em>over the aligned region</em>.</li>
</ul>
<p><strong>E-value</strong></p>
<ul>
<li>This is the number of hits equivalent to this hit that we would expect to see by chance alone.</li>
<li>Smaller E-values represent better hits, but an exact E-value cut off needs to be decided on a case by case basis.</li>
<li>E-values take into account the coverage and identity scores for each hit, and also the size of the database queried.</li>
</ul>
<p><strong>Bit score</strong> </p>
<ul>
<li>Similarly to E-values, bit scores summarise the sequence similarity between the query and database hit.</li>
<li>Bit scores are calculated independently from the query sequence length and the database size, as databases are constantly evolving this makes bit scores a constant statistical indicator.</li>
<li>A higher bit score indicates a better hit. </li>
</ul>
</div>
<hr>
<h2 id="examining-our-file">Examining our file<a class="headerlink" href="#examining-our-file" title="Permanent link">¶</a></h2>
<p>With this in mind lets look at the results from our <code>BLAST</code> job.</p>
<p>Return to your <code>blast_annotation/</code> folder, if you left it, and examine the new output file. Take a look at your results using the <code>less</code> or <code>head</code> command.</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>head<span class="w"> </span>output.txt
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>seq1    gi|1607238104|dbj|AP019558.1|   91.750  1794    148     0       1       1794    366818  3650250.0      2494    Mycoplasma bovis KG4397 DNA, complete genome    28903
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>seq1    gi|1441442372|gb|CP022588.1|    91.695  1794    149     0       1       1794    690905  6926980.0      2488    Mycoplasmopsis bovis strain MJ4 chromosome, complete genome     28903
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>seq1    gi|1315670167|emb|LT578453.1|   91.695  1794    149     0       1       1794    655989  6577820.0      2488    Mycoplasma bovis isolate JF4278 genome assembly, chromosome: I  28903
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>seq1    gi|2507795645|gb|CP058524.2|    91.695  1794    149     0       1       1794    401349  4031420.0      2488    Mycoplasmopsis bovis strain Mb49 chromosome     28903
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>seq1    gi|2507793515|gb|CP058496.2|    91.695  1794    149     0       1       1794    320995  3192020.0      2488    Mycoplasmopsis bovis strain Mb222 chromosome    28903
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>seq1    gi|2507792460|gb|CP058473.2|    91.695  1794    149     0       1       1794    1055619 10574120.0     2488    Mycoplasmopsis bovis strain VK22 chromosome     28903
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>seq1    gi|2507791648|gb|CP058453.2|    91.695  1794    149     0       1       1794    9966    11759 0.0      2488    Mycoplasmopsis bovis strain Mb287 chromosome    28903
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>seq1    gi|2507791648|gb|CP058453.2|    91.695  1794    149     0       1       1794    1120377 11221700.0     2488    Mycoplasmopsis bovis strain Mb287 chromosome    28903
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>seq1    gi|2506302979|gb|CP058448.2|    91.695  1794    149     0       1       1794    763992  7657850.0      2488    Mycoplasmopsis bovis strain Mb1 chromosome      28903
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a>seq1    gi|2506302187|gb|CP058432.2|    91.695  1794    149     0       1       1794    354705  3564980.0      2488    Mycoplasmopsis bovis strain Mb216 chromosome    28903
</code></pre></div>
</details>
</div>
<p>It looks like a table, but awkwardly there are no column names. However, the names of the columns correspond to the values we provided in our <code>slurm</code> script at the start of this session.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Column header</th>
<th style="text-align: left;">Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">qseqid</td>
<td style="text-align: left;">Sequence ID of the query sequence (input file)</td>
</tr>
<tr>
<td style="text-align: left;">sseqid</td>
<td style="text-align: left;">Sequence ID of the target sequence (reference database)</td>
</tr>
<tr>
<td style="text-align: left;">pident</td>
<td style="text-align: left;">Percentage of identical positions between query and target</td>
</tr>
<tr>
<td style="text-align: left;">length</td>
<td style="text-align: left;">Alignment length (sequence overlap) of the common region between query and target</td>
</tr>
<tr>
<td style="text-align: left;">mismatch</td>
<td style="text-align: left;">Number of mismatches between query and target</td>
</tr>
<tr>
<td style="text-align: left;">gapopen</td>
<td style="text-align: left;">Number of gap openings in the alignment</td>
</tr>
<tr>
<td style="text-align: left;">qstart</td>
<td style="text-align: left;">Position in the query sequence where alignment begins</td>
</tr>
<tr>
<td style="text-align: left;">qend</td>
<td style="text-align: left;">Position in the query sequence where alignment ends</td>
</tr>
<tr>
<td style="text-align: left;">sstart</td>
<td style="text-align: left;">Position in the target sequence where alignment begins</td>
</tr>
<tr>
<td style="text-align: left;">send</td>
<td style="text-align: left;">Position in the target sequence where alignment ends</td>
</tr>
<tr>
<td style="text-align: left;">evalue</td>
<td style="text-align: left;">The E-value for the query/target match, as described above</td>
</tr>
<tr>
<td style="text-align: left;">bitscore</td>
<td style="text-align: left;">The bit score for the query/target match, as described above</td>
</tr>
<tr>
<td style="text-align: left;">salltitles</td>
<td style="text-align: left;">Display All Subject Title(s) for the target sequence</td>
</tr>
<tr>
<td style="text-align: left;">staxids</td>
<td style="text-align: left;">Display the NCBI taxid value(s) for the target sequence</td>
</tr>
</tbody>
</table>
<p>Applying these to the layout, we get something more sensible:</p>
<details class="abstract">
<summary>Table layout</summary>
<table>
<thead>
<tr>
<th style="text-align: left;">qseqid</th>
<th style="text-align: left;">sseqid</th>
<th style="text-align: center;">pident</th>
<th style="text-align: center;">length</th>
<th style="text-align: center;">mismatch</th>
<th style="text-align: center;">gapopen</th>
<th style="text-align: center;">qstart</th>
<th style="text-align: center;">qend</th>
<th style="text-align: center;">sstart</th>
<th style="text-align: center;">send</th>
<th style="text-align: center;">evalue</th>
<th style="text-align: center;">bitscore</th>
<th style="text-align: left;">salltitles</th>
<th style="text-align: left;">staxids</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|1607238104|dbj|AP019558.1|</td>
<td style="text-align: center;">91.750</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">148</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">366818</td>
<td style="text-align: center;">365025</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2494</td>
<td style="text-align: left;">Mycoplasma bovis KG4397 DNA, complete genome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|1441442372|gb|CP022588.1|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">690905</td>
<td style="text-align: center;">692698</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain MJ4 chromosome, complete genome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|1315670167|emb|LT578453.1|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">655989</td>
<td style="text-align: center;">657782</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasma bovis isolate JF4278 genome assembly, chromosome: I</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2507795645|gb|CP058524.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">401349</td>
<td style="text-align: center;">403142</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain Mb49 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2507793515|gb|CP058496.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">320995</td>
<td style="text-align: center;">319202</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain Mb222 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2507792460|gb|CP058473.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">1055619</td>
<td style="text-align: center;">1057412</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain VK22 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2507791648|gb|CP058453.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">9966</td>
<td style="text-align: center;">11759</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain Mb287 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2507791648|gb|CP058453.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">1120377</td>
<td style="text-align: center;">1122170</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain Mb287 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2506302979|gb|CP058448.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">763992</td>
<td style="text-align: center;">765785</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain Mb1 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
<tr>
<td style="text-align: left;">seq1</td>
<td style="text-align: left;">gi|2506302187|gb|CP058432.2|</td>
<td style="text-align: center;">91.695</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1794</td>
<td style="text-align: center;">354705</td>
<td style="text-align: center;">356498</td>
<td style="text-align: center;">0.0</td>
<td style="text-align: center;">2488</td>
<td style="text-align: left;">Mycoplasmopsis bovis strain Mb216 chromosome</td>
<td style="text-align: left;">28903</td>
</tr>
</tbody>
</table>
</details>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>Download the <code>output.txt</code> file do your computer and open it in <code>Excel</code>. Look through the sequence annotations, and determine the most likely organism that each sequence was obtained from.</p>
<p>Are there are values in the results table that you are skeptical about? If so, raise them as a discussion with the group.</p>
<details class="circle-check">
<summary>Solution</summary>
<ul>
<li>seq1 = <em>Mycoplasma bovis</em></li>
<li>seq2 = <em>Bactrocera ritsemai</em></li>
<li>seq3 = Pepino mosaic virus</li>
<li>seq4 = No hit</li>
<li>seq5 = <em>Fusarium oxysporum</em> (reversed)</li>
</ul>
</details>
</div>
<hr>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../42_slurm_introduction/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 4.2 - Introduction to slurm">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                4.2 - Introduction to slurm
              </div>
            </div>
          </a>
        
        
          
          <a href="../../level2/11_shell_manipulation/" class="md-footer__link md-footer__link--next" aria-label="Next: 1.1 - Manipulating files in the shell">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                1.1 - Manipulating files in the shell
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Intermediate Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate", "navigation.tabs", "navigation.tabs.sticky"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>