<!DOCTYPE html><html lang="en" class="no-js"><head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="MPI, Genomics Aotearoa &amp; NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/hts_workshop_mpi/level1/32_nanopore_filtering/">
      
      
        <link rel="prev" href="../31_nanopore_inspection/">
      
      
        <link rel="next" href="../41_blastn_annotation/">
      
      
      <link rel="icon" href="../../theme_images/icon_fig.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.21">
    
    
      
        <title>3.2 - Filtering Nanopore data - High Throughput Sequencing</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2a3383ac.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&amp;display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  <link href="../../assets/stylesheets/glightbox.min.css" rel="stylesheet"><script src="../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#32-filtering-nanopore-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="High Throughput Sequencing" class="md-header__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            High Throughput Sequencing
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              3.2 - Filtering Nanopore data
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to light mode" type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to dark mode" type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="indigo" aria-label="Switch to system preference" type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"></path></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../11_nesi_connection/" class="md-tabs__link">
          
  
  
  Level1

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../level2/11_shell_manipulation/" class="md-tabs__link">
          
  
  
  Level2

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../proficiency/2023_proficiency/" class="md-tabs__link">
          
  
  
  Proficiency

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../supplementary/common_terms/" class="md-tabs__link">
          
  
  
  Supplementary

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="High Throughput Sequencing" class="md-nav__button md-logo" aria-label="High Throughput Sequencing" data-md-component="logo">
      
  <img src="../../theme_images/icon_fig.png" alt="logo">

    </a>
    High Throughput Sequencing
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/hts_workshop_mpi" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"></path></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/hts_workshop_mpi
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Level1
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Level1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../11_nesi_connection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Connecting to NeSI
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../12_shell_navigation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Introducing the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../13_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../21_illumina_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Assesing Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../22_illumina_filtering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Filtering Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../31_nanopore_inspection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Assesing Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    3.2 - Filtering Nanopore data
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#identifying-and-removing-adapters" class="md-nav__link">
    <span class="md-ellipsis">
      Identifying and removing adapters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#removing-short-and-low-quality-reads" class="md-nav__link">
    <span class="md-ellipsis">
      Removing short and low quality reads
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../41_blastn_annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Annotating with BLAST
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../42_slurm_introduction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Introduction to slurm
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../43_blast_interpretation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Interpreting BLAST results
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3">
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Level2
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Level2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/11_shell_manipulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.1 - Manipulating files in the shell
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/12_shell_variables/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.2 - Working with variables and loops
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/13_shell_redirection/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.3 - Working with redirection
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/14_nesi_environment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    1.4 - The NeSI computing environment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/21_assembly_de_novo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.1 - Genome Assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/22_assembly_spades/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.2 - Assembling short-read Illumina data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/23_assembly_flye/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Assembling Oxford Nanopore data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/24_assembly_evaluation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.3 - Evaluating the results of an assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/25_assembly_polishing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2.5 - Polishing of Oxford Nanopore assemblies
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/31_coverage_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.1 - Overview of mapping sequences to a reference genome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/32_illumina_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.2 - Mapping Illumina sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/33_nanopore_mapping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.3 - Mapping Oxford Nanopore sequences to a reference
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/34_mapping_filters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.4 - Filtering and compressing sam files
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/35_mapping_statistics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    3.5 - Summarising coverage information from a mapping file
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/41_annotation_overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.1 - Prediction of protein coding sequences
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/42_annotation_prodigal/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.2 - Prediction with prodigal
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/43_annotation_augustus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.3 - Prediction with AUGUSTUS
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/44_annotation_protein/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.4 - Protein annotation with BLAST-like methods
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../level2/45_annotation_kraken2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    4.5 - Classification of sequences with k-mer composition profiles
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4">
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Proficiency
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Proficiency
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 1)
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../proficiency/2023_proficiency_alternate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    2023 Level 2 proficiency testing (version 2)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5">
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Supplementary
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Supplementary
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/common_terms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Common terms
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/fastq_format/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    The FASTQ file format
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../supplementary/slurm_module_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Slurm module guide
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#identifying-and-removing-adapters" class="md-nav__link">
    <span class="md-ellipsis">
      Identifying and removing adapters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#removing-short-and-low-quality-reads" class="md-nav__link">
    <span class="md-ellipsis">
      Removing short and low quality reads
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="32-filtering-nanopore-data">3.2 - Filtering Nanopore data<a class="headerlink" href="#32-filtering-nanopore-data" title="Permanent link">¶</a></h1>
<div class="admonition clock">
<p class="admonition-title">time</p>
<ul>
<li>Teaching: 10 minutes</li>
<li>Exercises: 10 minutes</li>
</ul>
</div>
<div class="admonition circle-info">
<p class="admonition-title">Learning objectives</p>
<p><strong>Objectives</strong></p>
<ul>
<li>Be able to perform quality filtering of Nanopore data to remove short reads and adapter sequences</li>
</ul>
<p><strong>Key points</strong></p>
<ul>
<li>Sequences can be filtered to remove low quality reads and short reads using tools such as <code>Nanofilt</code></li>
<li>It is important to use tools designed specifically for long read data when performing assessment and trimming of Nanopore reads</li>
<li>Identifying and removing residual sequencing adapters is not always necessary, but tools exist for it if required</li>
</ul>
</div>
<hr>
<h2 id="identifying-and-removing-adapters">Identifying and removing adapters<a class="headerlink" href="#identifying-and-removing-adapters" title="Permanent link">¶</a></h2>
<p>Before we start worring about the raw sequencing quality of our reads, there is one factor to consider. When creating HTS libraries for sequencing, barcode and adpater sequences are appended to the ends of the template DNA to facilitate sequencing. These are an artificial construct, not reflective of the read biology of our sample, but are sequenced the same as the template DNA by the sequence device.</p>
<p>Adapters may be found at the start of end of a sequence, as with Illumina reads, but an issue unique to Nanopore sequencing is the occurence of <em>mid-sequence</em> adapters. This occurs when a sequence moves through the sequencing pore and is immediately followed by another read. If this happens quickly enough that the base calling tool can realise the first sequence has ended, the the new sequence is appended to tail of the first.</p>
<p>When this phenomenon happens you end up with a chimeric fusion of the first and second sequence, separated by the adapter of the second sequence. While mid-sequence adapaters artifacts are rare (typically less than 1% of a sequencing library) they are very easy to screen for, and it is important to identify and correct them in order to get the best classification or assembly result possible.</p>
<p>Like quality filtering, adapater removal (including mid-sequence adapters) can be performed during base calling, but if it was not, we can use the tool <code>porechop</code> to identify and remove adapaters:</p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>porechop<span class="w"> </span>--threads<span class="w"> </span><span class="m">2</span><span class="w"> </span>-i<span class="w"> </span>reads/nanopore.fq.gz<span class="w"> </span>-o<span class="w"> </span>results/nanopore.porechop.fq
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>Trimming<span class="w"> </span>adapters<span class="w"> </span>from<span class="w"> </span><span class="nb">read</span><span class="w"> </span>ends
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="w">     </span>SQK-NSK007_Y_Top:<span class="w"> </span>AATGTACTTCGTTCAGTTACGTATTGCT
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="w">  </span>SQK-NSK007_Y_Bottom:<span class="w"> </span>GCAATACGTAACTGAACGAAGT
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="w">     </span>1D2_part_2_start:<span class="w"> </span>CTTCGTTCAGTTACGTATTGCTGGCGTCTGCTT
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="w">       </span>1D2_part_2_end:<span class="w"> </span>CACCCAAGCAGACGCCAGCAATACGTAACT
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="m">3</span>,722<span class="w"> </span>/<span class="w"> </span><span class="m">3</span>,722<span class="w"> </span><span class="o">(</span><span class="m">100</span>.0%<span class="o">)</span>
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a>
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a><span class="m">3</span>,120<span class="w"> </span>/<span class="w"> </span><span class="m">3</span>,722<span class="w"> </span>reads<span class="w"> </span>had<span class="w"> </span>adapters<span class="w"> </span>trimmed<span class="w"> </span>from<span class="w"> </span>their<span class="w"> </span>start<span class="w"> </span><span class="o">(</span><span class="m">68</span>,630<span class="w"> </span>bp<span class="w"> </span>removed<span class="o">)</span>
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a><span class="w">  </span><span class="m">834</span><span class="w"> </span>/<span class="w"> </span><span class="m">3</span>,722<span class="w"> </span>reads<span class="w"> </span>had<span class="w"> </span>adapters<span class="w"> </span>trimmed<span class="w"> </span>from<span class="w"> </span>their<span class="w"> </span>end<span class="w"> </span><span class="o">(</span><span class="m">9</span>,680<span class="w"> </span>bp<span class="w"> </span>removed<span class="o">)</span>
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a>
<a id="__codelineno-1-12" name="__codelineno-1-12" href="#__codelineno-1-12"></a>
<a id="__codelineno-1-13" name="__codelineno-1-13" href="#__codelineno-1-13"></a>Splitting<span class="w"> </span>reads<span class="w"> </span>containing<span class="w"> </span>middle<span class="w"> </span>adapters
<a id="__codelineno-1-14" name="__codelineno-1-14" href="#__codelineno-1-14"></a><span class="m">3</span>,722<span class="w"> </span>/<span class="w"> </span><span class="m">3</span>,722<span class="w"> </span><span class="o">(</span><span class="m">100</span>.0%<span class="o">)</span>
<a id="__codelineno-1-15" name="__codelineno-1-15" href="#__codelineno-1-15"></a>
<a id="__codelineno-1-16" name="__codelineno-1-16" href="#__codelineno-1-16"></a><span class="m">20</span><span class="w"> </span>/<span class="w"> </span><span class="m">3</span>,722<span class="w"> </span>reads<span class="w"> </span>were<span class="w"> </span>split<span class="w"> </span>based<span class="w"> </span>on<span class="w"> </span>middle<span class="w"> </span>adapters
</code></pre></div>
</details>
</div>
<p>Watch the output as <code>porechop</code> runs, and see how many residual adapters were found in your mock data set. One of the really nice features of <code>porechop</code> is that we do not need to know which adapter sequences were used in library construction - the tool has a built-in reference set of the common adapter sequences and reads are screened for all by default.</p>
<div class="admonition question">
<p class="admonition-title">Exercise</p>
<p>From the output of <code>porechop</code>, answere the following questions:</p>
<ol>
<li>Which adapter sets were used to produce this library?</li>
<li>How many sequences were flagged as carrying a forward or reverse adapter sequence?</li>
<li>How many sequences had mid-sequence adapaters which needed removal?</li>
</ol>
<details class="circle-check">
<summary>Solution</summary>
<ol>
<li>The adapters from the <strong>SQK-NSK007</strong> and <strong>1D2</strong> sequencing kits were found in this library. </li>
<li>3,120 sequences were found to have a forward adapter attached, and 834 sequences had an end adapter attached.</li>
<li>20 of the reads were found ot have a mid-sequence adapter.</li>
</ol>
</details>
</div>
<hr>
<h2 id="removing-short-and-low-quality-reads">Removing short and low quality reads<a class="headerlink" href="#removing-short-and-low-quality-reads" title="Permanent link">¶</a></h2>
<p>Finally, we need to remove low quality data from the sequencing library. It can also be helpful to remove short reads from the data set at this point, as these can negatively impact the results of your downstream analysis in some situations.For example, if you are trying to assemble a genome from the data, the short reads may interfere with the assembly process and result in a worse genome that you would get if they were screened rom the library beforehand.</p>
<p><code>NanoFilt</code> is a tool that can filter reads by quality score and length. It is also capable of cropping a specified number of bases from the start or the end of a read, which can be useful sometimes as read quality is not uniform along the sequences - in Nanopore data it is common to see that the average quality score in the first few dozen nucleotides is significantly worse than the rest of the sequence.</p>
<p>We will not apply <code>NanoFilt</code> to:</p>
<ul>
<li>Remove all reads with quality scores under 15</li>
<li>Remove all reads shorter than 500 bp</li>
<li>Trim the first 50 nucleotides off all reads</li>
</ul>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>NanoFilt<span class="w"> </span>-q<span class="w"> </span><span class="m">15</span><span class="w"> </span>-l<span class="w"> </span><span class="m">500</span><span class="w"> </span>--headcrop<span class="w"> </span><span class="m">50</span><span class="w"> </span>&lt;<span class="w"> </span>results/nanopore.porechop.fq<span class="w"> </span>&gt;<span class="w"> </span>results/nanopore.qc.fq
</code></pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code>NanoFilt</code> behaves a bit strangely in the way it takes input data and writes output. Rather than specifying files with flags such as <code>-i</code> and <code>-o</code>, it reads a data from a stream called <code>stdin</code>, and writes to a separate channel called <code>stdout</code>.</p>
<p>Working with these channels is part of the Level 2 training, for now you just need to know that the <code>&lt;</code> character sets the <code>stdin</code> input, and <code>&gt;</code> specifies where the <code>stdout</code> data will go.</p>
</div>
<p>We can get a quick estimate for how much data passed our quality requirements by checking the number of sequences in the input and output files using Seqkit. Seqkit has a range of features but the <code>stats</code> command gives really good summary statistics for sequencing datasets. Here we are using the flag <code>-a</code> to return all the available stats. </p>
<div class="admonition terminal">
<p class="admonition-title">code</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>seqkit<span class="w"> </span>stats<span class="w"> </span>-a<span class="w"> </span>results/*.fq
</code></pre></div>
<details class="success">
<summary>Output</summary>
<table>
<thead>
<tr>
<th>file</th>
<th>format</th>
<th>type</th>
<th>num_seqs</th>
<th>sum_len</th>
<th>min_len</th>
<th>avg_len</th>
<th>max_len</th>
<th>Q1</th>
<th>Q2</th>
<th>Q3</th>
<th>sum_gap</th>
<th>N50</th>
<th>Q20(%)</th>
<th>Q30(%)</th>
<th>GC(%)</th>
</tr>
</thead>
<tbody>
<tr>
<td>results/nanopore.porechop.fq</td>
<td>FASTQ</td>
<td>DNA</td>
<td>3,735</td>
<td>45,214,457</td>
<td>6</td>
<td>12,105.6</td>
<td>82,753</td>
<td>2,723.5</td>
<td>8,489</td>
<td>17,927</td>
<td>0</td>
<td>21,118</td>
<td>69.79</td>
<td>50.97</td>
<td>30.83</td>
</tr>
<tr>
<td>results/nanopore.qc.fq</td>
<td>FASTQ</td>
<td>DNA</td>
<td>1,313</td>
<td>16,675,742</td>
<td>500</td>
<td>12,700.5</td>
<td>77,866</td>
<td>3,658</td>
<td>9,245</td>
<td>18,869</td>
<td>0</td>
<td>21,365</td>
<td>78.07</td>
<td>58.66</td>
<td>31.03</td>
</tr>
</tbody>
</table>
</details>
</div>
<p>How many sequences did we lose in the process?</p>
<hr>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer">
        
          
          <a href="../31_nanopore_inspection/" class="md-footer__link md-footer__link--prev" aria-label="Previous: 3.2 - Assesing Nanopore data">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                3.2 - Assesing Nanopore data
              </div>
            </div>
          </a>
        
        
          
          <a href="../41_blastn_annotation/" class="md-footer__link md-footer__link--next" aria-label="Next: 4.1 - Annotating with BLAST">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                4.1 - Annotating with BLAST
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Intermediate Shell for Bioinformatics workshop material is licensed under a <a rel="noopener" target="_blank" href="https://www.gnu.org/licenses/gpl-3.0.en.html">GNU General Public License v3.0</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand", "navigation.top", "navigation.footer", "content.code.copy", "content.code.annotate", "navigation.tabs", "navigation.tabs.sticky"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>